{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8aa228f6-1f9e-446b-bd3a-944ac913bc3e",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b56d31e2-34ca-480e-a32b-d70d6e6a3031",
   "metadata": {},
   "source": [
    "## nazwa plikow po blascie\n",
    "query_db_bl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7fdf9c57-bfb4-44c7-ac99-7a581de474fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from bioinfokit.analys import Fasta\n",
    "from Bio import SeqIO\n",
    "# helpers function\n",
    "\n",
    "def get_fasta(filename):\n",
    "    n = 0\n",
    "    sequence = \"\"\n",
    "    te_name = \"\"\n",
    "    with open(filename) as f:\n",
    "        for line in f.readlines():\n",
    "            if line.startswith('>'):\n",
    "                n = n + 1\n",
    "                if not n == 1:\n",
    "                    first_n = False\n",
    "                    yield te_name, sequence\n",
    "                te_name = line[1:].strip()\n",
    "                sequence = \"\"\n",
    "            else:\n",
    "                if te_name:\n",
    "                    sequence = sequence + line.strip()\n",
    "    yield te_name, sequence\n",
    "\n",
    "def get_len_fasta(filename):\n",
    "    counter = {}\n",
    "    with open(filename) as f:\n",
    "        te_name = \"\"\n",
    "        for line in f.readlines():\n",
    "            if line.startswith('>'):\n",
    "                te_name = line[1:].strip()\n",
    "                counter[te_name] = 0\n",
    "            else:\n",
    "                if te_name:\n",
    "                    counter[te_name] = counter[te_name] + len(line.strip()) - line.count(\"~\")\n",
    "    return counter\n",
    "\n",
    "def get_from_te(fasta, te_name):\n",
    "    te_len = get_len_fasta(fasta)[te_name]\n",
    "    return int(te_len - te_len * dif_percent / 100)\n",
    "\n",
    "def get_to_te(fasta, te_name):\n",
    "    te_len = get_len_fasta(fasta)[te_name]\n",
    "    return int(te_len + te_len * dif_percent / 100)\n",
    "\n",
    "def read_fasta(filename):\n",
    "    return pd.read_csv(filename, sep=\"\\t\", header=None).rename(headers_dict, axis=1)\n",
    "\n",
    "def save_fasta(df, filename):\n",
    "    df.to_csv(filename, sep='\\t', index=False, encoding='utf-8', header=False)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "fe2261af-9c56-4400-aeab-d929eb4d76c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nmask_genome(te_name, te_path, genome_path, True)\\ncorrect_first_part(False, True, True)\\noxi = correct_second_part()\\nsave_report(oxi)\\n#generate_second_report(\"report_Alex1__K10B_K10f\", \"doro\")\\n\\n'"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ont = [\"K10B\", \"K10f\", \"LK5_1_3\"]\n",
    "ont = [\"ccc\"]\n",
    "ont = [\"K10B\", \"K10f\"]\n",
    "ont = [\n",
    "    \"K10B\",\n",
    "    \"K10f\",\n",
    "    \"LK5_1_3\"\n",
    "]\n",
    "genome_path = \"genome/DCARv3.4.fa\"\n",
    "genome_path = \"genome/genomic2.fna\"\n",
    "te_path = \"te/Alex1_3.fas\"\n",
    "dif_percent = 30\n",
    "\n",
    "te_name = 'Alex1'\n",
    "te_len = get_len_fasta(te_path)[te_name]\n",
    "from_te = get_from_te(te_path, te_name)\n",
    "to_te = get_to_te(te_path, te_name)\n",
    "\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "mask_genome(te_name, te_path, genome_path, True)\n",
    "correct_first_part(False, True, True)\n",
    "oxi = correct_second_part()\n",
    "save_report(oxi)\n",
    "#generate_second_report(\"report_Alex1__K10B_K10f\", \"doro\")\n",
    "\n",
    "\"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3b790fca-e1b9-4207-bcb9-e4ace324321d",
   "metadata": {},
   "outputs": [],
   "source": [
    "headers = ['qseqid', 'sseqid',  'pident', 'length', 'mismatch', 'gapopen', 'qstart', 'qend', 'sstart', 'send', 'expect value', 'bitscore']\n",
    "headers_dict = {i[0]:i[1] for i in enumerate(headers)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "ecd1cc04-aba2-42fd-b345-a1cecd4f6aa6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "1\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "2\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "3\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "4\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "5\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "6\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "7\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "8\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "9\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "10\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "11\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "12\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "13\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "14\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "15\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "16\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "17\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "18\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "19\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "def rand_split(sequence, splits):\n",
    "    for splitLen in splits:\n",
    "        if splitLen > len(sequence):\n",
    "            break\n",
    "        yield sequence[:splitLen]\n",
    "        sequence = sequence[splitLen::]\n",
    "\n",
    "def rand_gen(int_list):\n",
    "    while True:\n",
    "        yield int(random.choice(int_list))\n",
    "\n",
    "def cut_genome(genome_path, ont_path, coverage, filename):\n",
    "    possibility_len = list(get_len_fasta(ont_path).values())\n",
    "    n = 0\n",
    "    for i in range(coverage):\n",
    "        print(i)\n",
    "        for name, genome_sequence in get_fasta(genome_path):\n",
    "            print(name)\n",
    "            with open(filename, 'a') as file:\n",
    "                for seq in rand_split(genome_sequence, rand_gen(possibility_len)):\n",
    "                    n = n + 1\n",
    "                    file.write(f\">{name}_{i}_{n}\\n{seq}\\n\")\n",
    "cut_genome(genome_path, \"ont/K10f.fasta\", 20, 'cutted_genome.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "674381f7-965c-4216-a694-5d30dda82d1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 05/14/2024 02:27:49\n",
      "New DB name:   /home/kiki/data2/genome/genome\n",
      "New DB title:  genome/genomic2.fna\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /home/kiki/data2/genome/genome\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 9 sequences in 6.73869 seconds.\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n"
     ]
    }
   ],
   "source": [
    "def mask_genome(te_name, te_path, genome_path, make_db=True):\n",
    "    os.system(f\"mkdir -p masked_genome\")\n",
    "    if make_db:\n",
    "        os.system(f\"makeblastdb -in {genome_path} -dbtype nucl -out genome/genome\")\n",
    "    os.system(f\"blastn -num_threads 20 -outfmt 6 -query {te_path} -db genome/genome -out masked_genome/{te_name}_genome.bl -perc_identity 0.9\")\n",
    "    te_genome_bl = read_fasta(f\"masked_genome/{te_name}_genome.bl\")\n",
    "    te_genome_bl = te_genome_bl[te_genome_bl.qseqid == te_name]\n",
    "    te_genome_bl[\"start\"] = te_genome_bl['sstart'].where(te_genome_bl['sstart'] <= te_genome_bl[\"send\"], other=te_genome_bl['send'])\n",
    "    te_genome_bl[\"stop\"] = te_genome_bl['sstart'].where(te_genome_bl['sstart'] > te_genome_bl[\"send\"], other=te_genome_bl['send'])\n",
    "    # todo\n",
    "    save_fasta(te_genome_bl[[\"sseqid\", \"start\", \"stop\"]], f\"masked_genome/to_mask_genome_{te_name}.txt\")\n",
    "    plus_te_genome_bl = te_genome_bl[te_genome_bl.sstart < te_genome_bl.send]\n",
    "    minus_te_genome_bl = te_genome_bl[te_genome_bl.sstart > te_genome_bl.send]\n",
    "\n",
    "    merge_headers = [\"sseqid\", \"sstart\", \"send\"]\n",
    "    save_fasta(plus_te_genome_bl[merge_headers].sort_values(by=merge_headers), f\"masked_genome/plus_to_mask_genome_{te_name}.bed\")\n",
    "    merge_headers = [\"sseqid\", \"send\", \"sstart\"]\n",
    "    save_fasta(minus_te_genome_bl[merge_headers].sort_values(by=merge_headers), f\"masked_genome/minus_to_mask_genome_{te_name}.bed\")\n",
    "\n",
    "    os.system(f\"bedtools merge -i masked_genome/plus_to_mask_genome_{te_name}.bed -d 200 > masked_genome/plus_to_mask_genome_{te_name}.bed.merged\")\n",
    "    os.system(f\"bedtools merge -i masked_genome/minus_to_mask_genome_{te_name}.bed -d 200 > masked_genome/minus_to_mask_genome_{te_name}.bed.merged\")\n",
    "    minus_merged =  pd.read_csv(f\"masked_genome/minus_to_mask_genome_{te_name}.bed.merged\", sep=\"\\t\", header=None)\n",
    "    plus_merged =  pd.read_csv(f\"masked_genome/plus_to_mask_genome_{te_name}.bed.merged\", sep=\"\\t\", header=None)\n",
    "    merged_genome = pd.concat(\n",
    "        [\n",
    "            minus_merged[\n",
    "                (minus_merged[2] - minus_merged[1] > from_te)\n",
    "                & (minus_merged[2] - minus_merged[1] < to_te)\n",
    "            ],\n",
    "            plus_merged[\n",
    "                (plus_merged[2] - plus_merged[1] > from_te)\n",
    "                & (plus_merged[2] - plus_merged[1] < to_te)\n",
    "            ],\n",
    "        ]\n",
    "    )\n",
    "    merged_genome[\"length\"] = merged_genome[2] - merged_genome[1]\n",
    "    save_fasta(merged_genome, f\"masked_genome/to_mask_genome_{te_name}.bed.merged\")\n",
    "    os.system(f\"bedtools maskfasta -fi {genome_path} -bed masked_genome/to_mask_genome_{te_name}.bed.merged -fo genome/{te_name}_masked_genome\")\n",
    "    return merged_genome\n",
    "masked_genome = mask_genome(te_name, te_path, genome_path, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "941b9f3e-faee-4f74-a83e-4d3f427f243d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chromosome2</td>\n",
       "      <td>30539632</td>\n",
       "      <td>30544357</td>\n",
       "      <td>4725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chromosome3</td>\n",
       "      <td>27717880</td>\n",
       "      <td>27722606</td>\n",
       "      <td>4726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chromosome5</td>\n",
       "      <td>8967214</td>\n",
       "      <td>8971939</td>\n",
       "      <td>4725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chromosome9</td>\n",
       "      <td>9220466</td>\n",
       "      <td>9226148</td>\n",
       "      <td>5682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chromosome9</td>\n",
       "      <td>28529339</td>\n",
       "      <td>28534568</td>\n",
       "      <td>5229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome1</td>\n",
       "      <td>18795334</td>\n",
       "      <td>18799058</td>\n",
       "      <td>3724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chromosome3</td>\n",
       "      <td>41159456</td>\n",
       "      <td>41164182</td>\n",
       "      <td>4726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chromosome4</td>\n",
       "      <td>35827649</td>\n",
       "      <td>35832374</td>\n",
       "      <td>4725</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0         1         2  length\n",
       "1   chromosome2  30539632  30544357    4725\n",
       "4   chromosome3  27717880  27722606    4726\n",
       "6   chromosome5   8967214   8971939    4725\n",
       "9   chromosome9   9220466   9226148    5682\n",
       "10  chromosome9  28529339  28534568    5229\n",
       "0   chromosome1  18795334  18799058    3724\n",
       "2   chromosome3  41159456  41164182    4726\n",
       "5   chromosome4  35827649  35832374    4725"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "masked_genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "b12184bd-e14f-4467-9277-d832ee5d0b03",
   "metadata": {},
   "outputs": [],
   "source": [
    "# first part\n",
    "def  create_subsequent_fasta(fasta, bed_df, file_base):\n",
    "    # create subsequents with left and right flanks\n",
    "    # site p or m\n",
    "    num = 4000\n",
    "\n",
    "    print(fasta)\n",
    "    records = SeqIO.parse(fasta, 'fasta')\n",
    "    record_dict = SeqIO.to_dict(records)\n",
    "    for bed_rec in  bed_df.itertuples():\n",
    "        seq_id = bed_rec.sseqid\n",
    "        record = record_dict[seq_id]\n",
    "        \n",
    "        te_start = bed_rec.sstart\n",
    "        te_end = bed_rec.send\n",
    "\n",
    "        left_start = te_start - num if te_start - num > 0 else 0\n",
    "        left_end = te_start\n",
    "\n",
    "        right_start = te_end\n",
    "        right_end = te_end + num if te_end + num < len(record) else len(record)\n",
    "\n",
    "        left_subseq_record = record[left_start:left_end]  # For the full record (with header)\n",
    "        right_subseq_record = record[right_start:right_end]\n",
    "\n",
    "        with open(f\"first/left_{file_base}\", 'a') as file:\n",
    "                file.write(f\">{record.id}\\n{left_subseq_record.seq}\\n\")\n",
    "        with open(f\"first/right_{file_base}\", 'a') as file:\n",
    "                file.write(f\">{record.id}\\n{right_subseq_record.seq}\\n\")\n",
    "\n",
    "def correct_first_part(fq=True, db=True, do_blast=True):\n",
    "    os.system(\"mkdir -p blast\")\n",
    "    os.system(\"mkdir -p first\")\n",
    "    if fq:\n",
    "        for file in ont:\n",
    "            os.system(f\"seqtk seq -a ont/{file}.fastq.gz > ont/{file}.fasta\")\n",
    "    if db:\n",
    "        print('okno')\n",
    "        for file in ont:\n",
    "            print(f\"makeblastdb -in ont/{file}.fasta -dbtype nucl -out ont/{file}\")\n",
    "            os.system(f\"makeblastdb -in ont/{file}.fasta -dbtype nucl -out ont/{file}\")\n",
    "    if do_blast:\n",
    "        print('okno2')\n",
    "        for file in ont:\n",
    "            print(\n",
    "                f\"blastn -num_threads 20 -outfmt 6 -query {te_path} -db ont/{file} \" \\\n",
    "                f\"-out blast/TE_{file}_{te_name}.bl -dust no -perc_identity 0.9\"\n",
    "            )\n",
    "            os.system(\n",
    "                f\"blastn -num_threads 20 -outfmt 6 -query {te_path} -db ont/{file} \" \\\n",
    "                f\"-out blast/TE_{file}_{te_name}.bl -dust no -perc_identity 0.9\"\n",
    "            )\n",
    "\n",
    "    for o in ont:\n",
    "        te_ont_bl = read_fasta(f\"blast/TE_{o}_{te_name}.bl\")\n",
    "        te_ont_bl = te_ont_bl[te_ont_bl.qseqid.str.contains(te_name)]\n",
    "        plus_te_ont_bl = te_ont_bl[te_ont_bl.sstart < te_ont_bl.send]\n",
    "        minus_te_ont_bl = te_ont_bl[te_ont_bl.sstart > te_ont_bl.send]\n",
    "    \n",
    "    \n",
    "        plus_te_ont_bed = plus_te_ont_bl[['sseqid', 'sstart', 'send', 'qseqid']].sort_values(by=['sseqid', 'sstart', 'send'])\n",
    "        minus_te_ont_bed = minus_te_ont_bl[['sseqid', 'send', 'sstart', 'qseqid']].sort_values(by=['sseqid', 'send', 'sstart'])\n",
    "    \n",
    "        save_fasta(minus_te_ont_bed, f'first/m_raw_first_blast_{o}_{te_name}.bed')\n",
    "        save_fasta(plus_te_ont_bed, f'first/p_raw_first_blast_{o}_{te_name}.bed')\n",
    "\n",
    "        os.system(f\"bedtools merge -i first/p_raw_first_blast_{o}_{te_name}.bed -d 1500 -c 4 -o collapse  > first/p_first_blast_{o}_{te_name}.bed.merged\")\n",
    "        os.system(f\"bedtools merge -i first/m_raw_first_blast_{o}_{te_name}.bed -d 1500 -c 4 -o collapse  > first/m_first_blast_{o}_{te_name}.bed.merged\")\n",
    "\n",
    "        print(\"first part\")\n",
    "    \n",
    "        plus_merged = pd.read_csv(\n",
    "            f\"first/p_first_blast_{o}_{te_name}.bed.merged\", sep=\"\\t\", header=None).rename(\n",
    "            {0:'sseqid', 1:'sstart', 2:'send', 3:'qseqid'}, axis=1\n",
    "        )\n",
    "        minus_merged = pd.read_csv(\n",
    "            f\"first/m_first_blast_{o}_{te_name}.bed.merged\", sep=\"\\t\", header=None).rename(\n",
    "            {0:'sseqid', 1:'sstart', 2:'send', 3:'qseqid'}, axis=1\n",
    "        )\n",
    "    \n",
    "        plus_filtered_bed = plus_merged[(plus_merged.send - plus_merged.sstart) > from_te]\n",
    "        minus_filtered_bed = minus_merged[(minus_merged.send - minus_merged.sstart) > from_te]\n",
    "\n",
    "        # todo in this place\n",
    "\n",
    "        save_fasta(pd.DataFrame({'id': plus_filtered_bed['sseqid'].unique()}), f'first/p_seq_list_{o}_{te_name}.txt')\n",
    "        save_fasta(pd.DataFrame({'id': minus_filtered_bed['sseqid'].unique()}), f'first/m_seq_list_{o}_{te_name}.txt')\n",
    "        #df.to_csv(filename, sep='\\t', index=False, encoding='utf-8', header=False)\n",
    "        print(\"second part\")\n",
    "    \n",
    "        #long\n",
    "        # it's filtered only our fasta\n",
    "        os.system(f\"seqtk subseq ont/{o}.fasta first/p_seq_list_{o}_{te_name}.txt > first/p_normal_{o}_{te_name}.fasta\")\n",
    "        os.system(f\"seqtk subseq ont/{o}.fasta first/m_seq_list_{o}_{te_name}.txt > first/m_normal_{o}_{te_name}.fasta\")\n",
    "        print(\"1 part\")\n",
    "\n",
    "        create_subsequent_fasta(f\"first/p_normal_{o}_{te_name}.fasta\", plus_filtered_bed,  f\"p_{o}_{te_name}.fasta\")\n",
    "        create_subsequent_fasta(f\"first/m_normal_{o}_{te_name}.fasta\", minus_filtered_bed,  f\"m_{o}_{te_name}.fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c1005ca-6ede-46f2-944b-305620683911",
   "metadata": {},
   "source": [
    "# Second part of analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0d39b878-4c49-4f21-a9a5-473f147b5c69",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def mid(records, row, ont_file):\n",
    "    ont_id = row[0]\n",
    "    for key, value in records.items():\n",
    "        if key[1] - row[2] < 1000 and key[1] - row[2] > -1000:\n",
    "            records[key][0] = records[key][0] + 1\n",
    "            records[key][1] =   f\"{records[key][1]},{ont_file}:{ont_id}\"\n",
    "            return records\n",
    "    records[row[1], row[2]] =  [1, f\"{ont_file}:{ont_id}\"]\n",
    "    return records\n",
    "\n",
    "def merge(report, records, ont_file):\n",
    "    #{chromosome: [start or end, number, ont_ids]\n",
    "    for record in records:\n",
    "        report = mid(report, record, ont_file)\n",
    "    return report\n",
    "\n",
    "    \n",
    "def merge_results(parts, ont_file, start, end):\n",
    "    start_te = []\n",
    "    end_te = []\n",
    "    for part in parts:\n",
    "        start_te.extend(part[0])\n",
    "        end_te.extend(part[1])\n",
    "    start_results = merge(start, start_te, ont_file)\n",
    "    end_results = merge(end, end_te, ont_file)\n",
    "    return start_results, end_results\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "6e06d225-049b-43e1-8dfe-599b5ad2b05d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#second part\n",
    "def blast_gen(query, db, out):\n",
    "    # i need only 1 target\n",
    "    os.system(f\"blastn -num_threads 20 -outfmt 6 -query {query} -db {db} -out {out} -perc_identity 0.9 -max_target_seqs 1\")\n",
    "    return pd.read_csv(out, sep=\"\\t\", header=None).rename(headers_dict, axis=1)\n",
    "\n",
    "def _get_ends(df, field_name):\n",
    "    records = []\n",
    "    unique_names = set(df.qseqid.unique())\n",
    "    for name in unique_names:\n",
    "        df_record = df[df.qseqid == name]\n",
    "        if len(df_record):\n",
    "            record = df_record.iloc[0]\n",
    "            records.append(\n",
    "                (\n",
    "                    record['qseqid'],\n",
    "                    record['sseqid'],\n",
    "                    record[field_name],\n",
    "                )\n",
    "            )\n",
    "    return records\n",
    "        \n",
    "def get_results(left_blast, right_blast):\n",
    "    te_start = []\n",
    "    te_end = []\n",
    "\n",
    "    plus_left = left_blast[left_blast.sstart <= left_blast.send]\n",
    "    plus_left[\"strain\"] = \"+\"\n",
    "    plus_right =  right_blast[right_blast.sstart <= right_blast.send]\n",
    "    plus_right[\"strain\"] = \"+\"\n",
    "    \n",
    "    minus_left = left_blast[left_blast.sstart > left_blast.send]\n",
    "    minus_left[\"strain\"] = \"-\"\n",
    "    minus_right = right_blast[right_blast.sstart > right_blast.send]\n",
    "    minus_right[\"strain\"] = \"-\"\n",
    "\n",
    "\n",
    "    filtered_plus_left =  plus_left[plus_left.length > 600]\n",
    "    filtered_minus_right =  minus_right[minus_right.length > 600]\n",
    "\n",
    "    filtered_plus_right =   plus_right[plus_right.length > 600]\n",
    "    filtered_minus_left = minus_left[minus_left.length > 600]\n",
    "\n",
    "    # start te\n",
    "    te_start.extend(_get_ends(filtered_plus_left, \"send\"))\n",
    "    te_start.extend(_get_ends(filtered_minus_right, \"send\"))\n",
    "    \n",
    "    # end te\n",
    "    te_end.extend(_get_ends(filtered_plus_right, \"sstart\"))\n",
    "    te_end.extend(_get_ends(filtered_minus_left, \"sstart\"))\n",
    "\n",
    "    return (te_start, te_end)\n",
    "\n",
    "\n",
    "def correct_second_part(blast_db=True):\n",
    "    report = {}\n",
    "    os.system(\"mkdir -p final\")\n",
    "    os.system(\"mkdir -p second\")\n",
    "    pd.options.mode.copy_on_write = True\n",
    "    if blast_db:\n",
    "        # todo\n",
    "        os.system(f\"makeblastdb -in {genome_path}  -dbtype nucl -out genome/masked_genome\")\n",
    "    for o in ont:\n",
    "        l_m = blast_gen(\n",
    "            f\"first/left_m_{o}_{te_name}.fasta\",\n",
    "            f\"genome/masked_genome\",\n",
    "            f\"second/left_m_{o}_{te_name}_genome.bl\",\n",
    "        ).sort_values(['qseqid', 'bitscore'], ascending=False).groupby(['qseqid'], as_index=False).first()\n",
    "        r_m = blast_gen(\n",
    "            f\"first/right_m_{o}_{te_name}.fasta\",\n",
    "            f\"genome/masked_genome\",\n",
    "            f\"second/right_m_{o}_{te_name}_genome.bl\",\n",
    "        ).sort_values(['qseqid', 'bitscore'], ascending=False).groupby(['qseqid'], as_index=False).first()\n",
    "        l_p = blast_gen(\n",
    "            f\"first/left_p_{o}_{te_name}.fasta\",\n",
    "            f\"genome/masked_genome\",\n",
    "            f\"second/left_p_{o}_{te_name}_genome.bl\",\n",
    "        ).sort_values(['qseqid', 'bitscore'], ascending=False).groupby(['qseqid'], as_index=False).first()\n",
    "        r_p = blast_gen(\n",
    "            f\"first/right_p_{o}_{te_name}.fasta\",\n",
    "            f\"genome/masked_genome\",\n",
    "            f\"second/right_p_{o}_{te_name}_genome.bl\",\n",
    "        ).sort_values(['qseqid', 'bitscore'], ascending=False).groupby(['qseqid'], as_index=False).first()\n",
    "\n",
    "        m_report = get_results(l_m, r_m)\n",
    "        p_report = get_results(l_p, r_p)\n",
    "\n",
    "        # todo\n",
    "        if report:\n",
    "            report = merge_results([m_report, p_report], o, report[0], report[1])\n",
    "        else:\n",
    "            report = merge_results([m_report, p_report], o, {}, {})\n",
    "    return report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "72cd00e7-b3a3-450e-91b6-0e73dfe431cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import copy\n",
    "def get_files():\n",
    "    files = \"\"\n",
    "    for o in ont:\n",
    "        files = files + f\"_{o}\"\n",
    "    return files\n",
    "\n",
    "\n",
    "def save_report(data):\n",
    "    start_c = copy.deepcopy(data[0])\n",
    "    end_c = copy.deepcopy(data[1])\n",
    "    report = {}\n",
    "    def get_string(start_te, end_te):\n",
    "        return end_te - start_te\n",
    "        return 'reference' if end_te - start_te < 200 else end_te - start_te\n",
    "    \n",
    "    for end_key, end_value in data[1].items():\n",
    "        #del end_c[end_key]\n",
    "        for start_key, start_value in data[0].items():\n",
    "            if end_key[0] == start_key[0]:\n",
    "                if end_key[1] - start_key[1] < 200 and end_key[1] - start_key[1] > -20 :\n",
    "                    #reference\n",
    "                    if end_c.get(end_key):\n",
    "                        del end_c[end_key]\n",
    "                    if start_c.get(start_key):\n",
    "                        del start_c[start_key]\n",
    "                    report[start_key[0], start_key[1], end_key[1]] = start_value + end_value\n",
    "                if end_key[1] - start_key[1] > from_te and end_key[1] - start_key[1] < to_te:\n",
    "                    #non reference\n",
    "                    if end_c.get(end_key):\n",
    "                        del end_c[end_key]\n",
    "                    if start_c.get(start_key):\n",
    "                        del start_c[start_key]\n",
    "                    report[start_key[0], start_key[1], end_key[1]] = start_value + end_value\n",
    "\n",
    "    with open(f\"report_{te_name}_{get_files()}\", \"w\") as file:\n",
    "        for key, value in report.items():\n",
    "                file.write(f\"{key[0]}\\t{key[1]}\\t{key[2]}\\t{get_string(key[1], key[2])}\\t{value[0]}\\t{value[1]}\\t{value[2]}\\t{value[3]}\\n\")\n",
    "        for key, value in start_c.items():\n",
    "                file.write(f\"{key[0]}\\t{key[1]}\\tNA\\tUNCOMPLETE\\t{value[0]}\\t{value[1]}\\tNA\\tNA\\n\")\n",
    "        for key, value in end_c.items():\n",
    "                file.write(f\"{key[0]}\\tNA\\t{key[1]}\\tNA\\tUNCOMPLETE\\tNA\\t{value[0]}\\t{value[1]}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "59a2eff5-6cca-4787-a5df-7f7fcd43f203",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_num_and_rec(line, reference):\n",
    "    start_flag = False\n",
    "    stop_flag = False\n",
    "    start_set = set()\n",
    "    stop_set = set()\n",
    "    result = {}\n",
    "    data = line.split('\\t')\n",
    "    result[\"chromosom\"] = data[0]\n",
    "    result[\"start\"] = data[1]\n",
    "    result[\"stop\"] = data[2]\n",
    "    result[\"length\"] = data[3]\n",
    "    result[\"pkt\"] = 4\n",
    "\n",
    "    is_reference = reference[\n",
    "        reference[0] == result[\"chromosom\"]\n",
    "    ]\n",
    "    #print(len(is_reference))\n",
    "\n",
    "    try:\n",
    "        st = int(data[1])\n",
    "    except:\n",
    "        st = 100000000\n",
    "    try:\n",
    "        sp = int(data[2])\n",
    "    except:\n",
    "        sp = 100000000\n",
    "    ref = reference[\n",
    "        (reference[0] == result[\"chromosom\"])\n",
    "        & (abs(reference[1] - st) < 1000)\n",
    "        & (abs(reference[2] - sp) < 1000)\n",
    "    ]\n",
    "    if len(ref):\n",
    "        result[\"is_reference\"] = \"ref\"\n",
    "    elif abs(st - sp) < 1000:\n",
    "        result[\"is_reference\"] = \"non_ref\"\n",
    "    else:\n",
    "        result[\"is_reference\"] = \"unnown\"\n",
    "    \n",
    "    for o in ont:\n",
    "        result[f\"num_start_{o}\"] = 0\n",
    "        result[f\"rec_start_{o}\"] = \"\"\n",
    "        result[f\"num_stop_{o}\"] = 0\n",
    "        result[f\"rec_stop_{o}\"] = \"\"\n",
    "    try:\n",
    "        start = data[5]\n",
    "        stop = data[7]\n",
    "    except:\n",
    "        return result\n",
    "    records_start = start.split(',')\n",
    "    records_stop = stop.split(',')\n",
    "    try:\n",
    "        for record in records_start:\n",
    "            start_flag = True\n",
    "            r = record.split(\":\")\n",
    "            num_res = result.get(f\"num_start_{r[0]}\")\n",
    "            rec_res = result.get(f\"rec_start_{r[0]}\")\n",
    "            result[f\"num_start_{r[0]}\"] = num_res + 1\n",
    "            result[f\"rec_start_{r[0]}\"] = rec_res + r[1] if not rec_res else rec_res + ',' + r[1]\n",
    "            start_set.add(r[1])\n",
    "    except:\n",
    "        pass\n",
    "    try:\n",
    "        for record in records_stop:\n",
    "            stop_flag = True\n",
    "            r = record.split(\":\")\n",
    "            num_res = result.get(f\"num_stop_{r[0]}\")\n",
    "            rec_res = result.get(f\"rec_stop_{r[0]}\")\n",
    "            result[f\"num_stop_{r[0]}\"] = num_res + 1\n",
    "            result[f\"rec_stop_{r[0]}\"] = rec_res + r[1] if not rec_res else rec_res + ',' + r[1]\n",
    "            stop_set.add(r[1])\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "    if start_flag and stop_flag:\n",
    "        result[\"pkt\"] = 3\n",
    "    for o in ont:\n",
    "        if  result[f\"num_start_{o}\"] and result[f\"num_stop_{o}\"]:\n",
    "            result[\"pkt\"] = 2\n",
    "    if start_set.intersection(stop_set):\n",
    "        result[\"pkt\"] = 1\n",
    "    return result\n",
    "\n",
    "def generate_second_report(filename_in, filename_out, reference):\n",
    "    num_rec_header =  \"chromosom\\tstart\\tstop\\tlength\\tpkt\"\n",
    "    num_rec_header =  \"pkt\\tchromosom\\tstart\\tstop\\tlength\\tis_ref\"\n",
    "    num_rec =  \"{chromosom}\\t{start}\\t{stop}\\t{length}\\t{pkt}\"\n",
    "    num_rec =  \"{pkt}\\t{chromosom}\\t{start}\\t{stop}\\t{length}\\t{is_reference}\"\n",
    "    for o in ont:\n",
    "        num_rec_header = num_rec_header +  f\"\\tnum_start_{o}\\trec_start_{o}\\tnum_stop_{o}\\trec_stop_{o}\"\n",
    "        num_rec = num_rec +  f\"\\t{{num_start_{o}}}\\t{{rec_start_{o}}}\\t{{num_stop_{o}}}\\t{{rec_stop_{o}}}\"\n",
    "    with open(filename_out, \"w\") as f_out:\n",
    "            f_out.write(f\"{num_rec_header}\\n\")\n",
    "    with open(filename_in) as f_in:\n",
    "        with open(filename_out, \"a\") as f_out:\n",
    "            for line in f_in.readlines():\n",
    "                f_out.write(\n",
    "                    f\"{num_rec.format(**get_num_and_rec(line, reference))}\\n\"\n",
    "                )\n",
    "generate_second_report(\"report_Alex1__K10B_K10f_LK5_1_3_old_genome\", \"old_report\", masked_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "2d636a75-657e-4ff6-a48f-c7b07b5aca1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#oxi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "837bca68-6433-4961-9c08-b624c9ec3c83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chromosome2</td>\n",
       "      <td>30539632</td>\n",
       "      <td>30544357</td>\n",
       "      <td>4725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chromosome3</td>\n",
       "      <td>27717880</td>\n",
       "      <td>27722606</td>\n",
       "      <td>4726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chromosome5</td>\n",
       "      <td>8967214</td>\n",
       "      <td>8971939</td>\n",
       "      <td>4725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chromosome9</td>\n",
       "      <td>9220466</td>\n",
       "      <td>9226148</td>\n",
       "      <td>5682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chromosome9</td>\n",
       "      <td>28529339</td>\n",
       "      <td>28534568</td>\n",
       "      <td>5229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome1</td>\n",
       "      <td>18795334</td>\n",
       "      <td>18799058</td>\n",
       "      <td>3724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chromosome3</td>\n",
       "      <td>41159456</td>\n",
       "      <td>41164182</td>\n",
       "      <td>4726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chromosome4</td>\n",
       "      <td>35827649</td>\n",
       "      <td>35832374</td>\n",
       "      <td>4725</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0         1         2  length\n",
       "1   chromosome2  30539632  30544357    4725\n",
       "4   chromosome3  27717880  27722606    4726\n",
       "6   chromosome5   8967214   8971939    4725\n",
       "9   chromosome9   9220466   9226148    5682\n",
       "10  chromosome9  28529339  28534568    5229\n",
       "0   chromosome1  18795334  18799058    3724\n",
       "2   chromosome3  41159456  41164182    4726\n",
       "5   chromosome4  35827649  35832374    4725"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "masked_genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9035e5ac-85c9-4e43-9640-f36fd6a0fce5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
