{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8aa228f6-1f9e-446b-bd3a-944ac913bc3e",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b56d31e2-34ca-480e-a32b-d70d6e6a3031",
   "metadata": {},
   "source": [
    "## nazwa plikow po blascie\n",
    "query_db_bl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7fdf9c57-bfb4-44c7-ac99-7a581de474fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from bioinfokit.analys import Fasta\n",
    "from Bio import SeqIO\n",
    "# helpers function\n",
    "\n",
    "def get_fasta(filename):\n",
    "    n = 0\n",
    "    sequence = \"\"\n",
    "    te_name = \"\"\n",
    "    with open(filename) as f:\n",
    "        for line in f.readlines():\n",
    "            if line.startswith('>'):\n",
    "                n = n + 1\n",
    "                if not n == 1:\n",
    "                    first_n = False\n",
    "                    yield te_name, sequence\n",
    "                te_name = line[1:].strip()\n",
    "                sequence = \"\"\n",
    "            else:\n",
    "                if te_name:\n",
    "                    sequence = sequence + line.strip()\n",
    "    yield te_name, sequence\n",
    "\n",
    "def get_len_fasta(filename):\n",
    "    counter = {}\n",
    "    with open(filename) as f:\n",
    "        te_name = \"\"\n",
    "        for line in f.readlines():\n",
    "            if line.startswith('>'):\n",
    "                te_name = line[1:].strip()\n",
    "                counter[te_name] = 0\n",
    "            else:\n",
    "                if te_name:\n",
    "                    counter[te_name] = counter[te_name] + len(line.strip()) - line.count(\"~\")\n",
    "    return counter\n",
    "\n",
    "def get_from_te(fasta, te_name):\n",
    "    te_len = get_len_fasta(fasta)[te_name]\n",
    "    return int(te_len - te_len * dif_percent / 100)\n",
    "\n",
    "def get_to_te(fasta, te_name):\n",
    "    te_len = get_len_fasta(fasta)[te_name]\n",
    "    return int(te_len + te_len * dif_percent / 100)\n",
    "\n",
    "def read_fasta(filename):\n",
    "    return pd.read_csv(filename, sep=\"\\t\", header=None).rename(headers_dict, axis=1)\n",
    "\n",
    "def save_fasta(df, filename):\n",
    "    df.to_csv(filename, sep='\\t', index=False, encoding='utf-8', header=False)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fe2261af-9c56-4400-aeab-d929eb4d76c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 05/21/2024 09:30:28\n",
      "New DB name:   /home/kiki/data2/genome/genome\n",
      "New DB title:  genome/genomic2.fna\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /home/kiki/data2/genome/genome\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 9 sequences in 4.54529 seconds.\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "okno\n",
      "makeblastdb -in ont/K10B.fasta -dbtype nucl -out ont/K10B\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/21/2024 09:30:33\n",
      "New DB name:   /home/kiki/data2/ont/K10B\n",
      "New DB title:  ont/K10B.fasta\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /home/kiki/data2/ont/K10B\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 6939343 sequences in 234.7 seconds.\n",
      "\n",
      "\n",
      "makeblastdb -in ont/K10f.fasta -dbtype nucl -out ont/K10f\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/21/2024 09:34:29\n",
      "New DB name:   /home/kiki/data2/ont/K10f\n",
      "New DB title:  ont/K10f.fasta\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /home/kiki/data2/ont/K10f\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 4775182 sequences in 160.619 seconds.\n",
      "\n",
      "\n",
      "makeblastdb -in ont/LK5_1_3.fasta -dbtype nucl -out ont/LK5_1_3\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/21/2024 09:37:11\n",
      "New DB name:   /home/kiki/data2/ont/LK5_1_3\n",
      "New DB title:  ont/LK5_1_3.fasta\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /home/kiki/data2/ont/LK5_1_3\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 1393171 sequences in 133.314 seconds.\n",
      "\n",
      "\n",
      "okno2\n",
      "blastn -num_threads 20 -outfmt 6 -query te/Alex1_3.fas -db ont/K10B -out blast/TE_K10B_Alex1.bl -dust no -perc_identity 0.9\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blastn -num_threads 20 -outfmt 6 -query te/Alex1_3.fas -db ont/K10f -out blast/TE_K10f_Alex1.bl -dust no -perc_identity 0.9\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blastn -num_threads 20 -outfmt 6 -query te/Alex1_3.fas -db ont/LK5_1_3 -out blast/TE_LK5_1_3_Alex1.bl -dust no -perc_identity 0.9\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "first part\n",
      "second part\n",
      "1 part\n",
      "first/p_normal_K10B_Alex1.fasta\n",
      "first/m_normal_K10B_Alex1.fasta\n",
      "first part\n",
      "second part\n",
      "1 part\n",
      "first/p_normal_K10f_Alex1.fasta\n",
      "first/m_normal_K10f_Alex1.fasta\n",
      "first part\n",
      "second part\n",
      "1 part\n",
      "first/p_normal_LK5_1_3_Alex1.fasta\n",
      "first/m_normal_LK5_1_3_Alex1.fasta\n"
     ]
    }
   ],
   "source": [
    "ont = [\"K10B\", \"K10f\", \"LK5_1_3\"]\n",
    "ont = [\"K10B\", \"K10f\"]\n",
    "ont = [\"ccc\"]\n",
    "ont = [\n",
    "    \"K10B\",\n",
    "    \"K10f\",\n",
    "    \"LK5_1_3\"\n",
    "]\n",
    "genome_path = \"genome/DCARv3.4.fa\"\n",
    "genome_path = \"genome/genomic2.fna\"\n",
    "te_path = \"te/Alex1_3.fas\"\n",
    "dif_percent = 30\n",
    "\n",
    "te_name = 'Alex1'\n",
    "te_len = get_len_fasta(te_path)[te_name]\n",
    "from_te = get_from_te(te_path, te_name)\n",
    "to_te = get_to_te(te_path, te_name)\n",
    "\n",
    "\n",
    "\n",
    "mask_genome(te_name, te_path, genome_path, True)\n",
    "correct_first_part(False, True, True)\n",
    "#oxi = correct_second_part()\n",
    "#save_report(oxi)\n",
    "#generate_second_report(\"report_Alex1__K10B_K10f\", \"doro\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3b790fca-e1b9-4207-bcb9-e4ace324321d",
   "metadata": {},
   "outputs": [],
   "source": [
    "headers = ['qseqid', 'sseqid',  'pident', 'length', 'mismatch', 'gapopen', 'qstart', 'qend', 'sstart', 'send', 'expect value', 'bitscore']\n",
    "headers_dict = {i[0]:i[1] for i in enumerate(headers)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ecd1cc04-aba2-42fd-b345-a1cecd4f6aa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "def rand_split(sequence, splits):\n",
    "    for splitLen in splits:\n",
    "        if splitLen > len(sequence):\n",
    "            break\n",
    "        yield sequence[:splitLen]\n",
    "        sequence = sequence[splitLen::]\n",
    "\n",
    "def rand_gen(int_list):\n",
    "    while True:\n",
    "        yield int(random.choice(int_list))\n",
    "\n",
    "def cut_genome(genome_path, ont_path, coverage, filename):\n",
    "    possibility_len = list(get_len_fasta(ont_path).values())\n",
    "    n = 0\n",
    "    for i in range(coverage):\n",
    "        print(i)\n",
    "        for name, genome_sequence in get_fasta(genome_path):\n",
    "            print(name)\n",
    "            with open(filename, 'a') as file:\n",
    "                for seq in rand_split(genome_sequence, rand_gen(possibility_len)):\n",
    "                    n = n + 1\n",
    "                    file.write(f\">{name}_{i}_{n}\\n{seq}\\n\")\n",
    "#cut_genome(genome_path, \"ont/K10f.fasta\", 20, 'cutted_genome.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "674381f7-965c-4216-a694-5d30dda82d1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 05/21/2024 09:29:57\n",
      "New DB name:   /home/kiki/data2/genome/genome\n",
      "New DB title:  genome/genomic2.fna\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /home/kiki/data2/genome/genome\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 9 sequences in 4.27525 seconds.\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n"
     ]
    }
   ],
   "source": [
    "def mask_genome(te_name, te_path, genome_path, make_db=True):\n",
    "    os.system(f\"mkdir -p masked_genome\")\n",
    "    if make_db:\n",
    "        os.system(f\"makeblastdb -in {genome_path} -dbtype nucl -out genome/genome\")\n",
    "    os.system(f\"blastn -num_threads 20 -outfmt 6 -query {te_path} -db genome/genome -out masked_genome/{te_name}_genome.bl -perc_identity 0.9\")\n",
    "    te_genome_bl = read_fasta(f\"masked_genome/{te_name}_genome.bl\")\n",
    "    te_genome_bl = te_genome_bl[te_genome_bl.qseqid == te_name]\n",
    "    te_genome_bl[\"start\"] = te_genome_bl['sstart'].where(te_genome_bl['sstart'] <= te_genome_bl[\"send\"], other=te_genome_bl['send'])\n",
    "    te_genome_bl[\"stop\"] = te_genome_bl['sstart'].where(te_genome_bl['sstart'] > te_genome_bl[\"send\"], other=te_genome_bl['send'])\n",
    "    # todo\n",
    "    save_fasta(te_genome_bl[[\"sseqid\", \"start\", \"stop\"]], f\"masked_genome/to_mask_genome_{te_name}.txt\")\n",
    "    plus_te_genome_bl = te_genome_bl[te_genome_bl.sstart < te_genome_bl.send]\n",
    "    minus_te_genome_bl = te_genome_bl[te_genome_bl.sstart > te_genome_bl.send]\n",
    "\n",
    "    merge_headers = [\"sseqid\", \"sstart\", \"send\"]\n",
    "    save_fasta(plus_te_genome_bl[merge_headers].sort_values(by=merge_headers), f\"masked_genome/plus_to_mask_genome_{te_name}.bed\")\n",
    "    merge_headers = [\"sseqid\", \"send\", \"sstart\"]\n",
    "    save_fasta(minus_te_genome_bl[merge_headers].sort_values(by=merge_headers), f\"masked_genome/minus_to_mask_genome_{te_name}.bed\")\n",
    "\n",
    "    os.system(f\"bedtools merge -i masked_genome/plus_to_mask_genome_{te_name}.bed -d 200 > masked_genome/plus_to_mask_genome_{te_name}.bed.merged\")\n",
    "    os.system(f\"bedtools merge -i masked_genome/minus_to_mask_genome_{te_name}.bed -d 200 > masked_genome/minus_to_mask_genome_{te_name}.bed.merged\")\n",
    "    minus_merged =  pd.read_csv(f\"masked_genome/minus_to_mask_genome_{te_name}.bed.merged\", sep=\"\\t\", header=None)\n",
    "    plus_merged =  pd.read_csv(f\"masked_genome/plus_to_mask_genome_{te_name}.bed.merged\", sep=\"\\t\", header=None)\n",
    "    merged_genome = pd.concat(\n",
    "        [\n",
    "            minus_merged[\n",
    "                (minus_merged[2] - minus_merged[1] > from_te)\n",
    "                & (minus_merged[2] - minus_merged[1] < to_te)\n",
    "            ],\n",
    "            plus_merged[\n",
    "                (plus_merged[2] - plus_merged[1] > from_te)\n",
    "                & (plus_merged[2] - plus_merged[1] < to_te)\n",
    "            ],\n",
    "        ]\n",
    "    )\n",
    "    merged_genome[\"length\"] = merged_genome[2] - merged_genome[1]\n",
    "    save_fasta(merged_genome, f\"masked_genome/to_mask_genome_{te_name}.bed.merged\")\n",
    "    os.system(f\"bedtools maskfasta -fi {genome_path} -bed masked_genome/to_mask_genome_{te_name}.bed.merged -fo genome/{te_name}_masked_genome\")\n",
    "    return merged_genome\n",
    "masked_genome = mask_genome(te_name, te_path, genome_path, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "941b9f3e-faee-4f74-a83e-4d3f427f243d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chromosome2</td>\n",
       "      <td>30539632</td>\n",
       "      <td>30544357</td>\n",
       "      <td>4725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chromosome3</td>\n",
       "      <td>27717880</td>\n",
       "      <td>27722606</td>\n",
       "      <td>4726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chromosome5</td>\n",
       "      <td>8967214</td>\n",
       "      <td>8971939</td>\n",
       "      <td>4725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chromosome9</td>\n",
       "      <td>9220466</td>\n",
       "      <td>9226148</td>\n",
       "      <td>5682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chromosome9</td>\n",
       "      <td>28529339</td>\n",
       "      <td>28534568</td>\n",
       "      <td>5229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome1</td>\n",
       "      <td>18795334</td>\n",
       "      <td>18799058</td>\n",
       "      <td>3724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chromosome3</td>\n",
       "      <td>41159456</td>\n",
       "      <td>41164182</td>\n",
       "      <td>4726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chromosome4</td>\n",
       "      <td>35827649</td>\n",
       "      <td>35832374</td>\n",
       "      <td>4725</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0         1         2  length\n",
       "1   chromosome2  30539632  30544357    4725\n",
       "4   chromosome3  27717880  27722606    4726\n",
       "6   chromosome5   8967214   8971939    4725\n",
       "9   chromosome9   9220466   9226148    5682\n",
       "10  chromosome9  28529339  28534568    5229\n",
       "0   chromosome1  18795334  18799058    3724\n",
       "2   chromosome3  41159456  41164182    4726\n",
       "5   chromosome4  35827649  35832374    4725"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "masked_genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b12184bd-e14f-4467-9277-d832ee5d0b03",
   "metadata": {},
   "outputs": [],
   "source": [
    "# first part\n",
    "def  create_subsequent_fasta(fasta, bed_df, file_base):\n",
    "    # create subsequents with left and right flanks\n",
    "    # site p or m\n",
    "    num = 4000\n",
    "\n",
    "    print(fasta)\n",
    "    records = SeqIO.parse(fasta, 'fasta')\n",
    "    record_dict = SeqIO.to_dict(records)\n",
    "    for bed_rec in  bed_df.itertuples():\n",
    "        seq_id = bed_rec.sseqid\n",
    "        record = record_dict[seq_id]\n",
    "        \n",
    "        te_start = bed_rec.sstart\n",
    "        te_end = bed_rec.send\n",
    "\n",
    "        left_start = te_start - num if te_start - num > 0 else 0\n",
    "        left_end = te_start\n",
    "\n",
    "        right_start = te_end\n",
    "        right_end = te_end + num if te_end + num < len(record) else len(record)\n",
    "\n",
    "        left_subseq_record = record[left_start:left_end]  # For the full record (with header)\n",
    "        right_subseq_record = record[right_start:right_end]\n",
    "\n",
    "        with open(f\"first/left_{file_base}\", 'a') as file:\n",
    "                file.write(f\">{record.id}\\n{left_subseq_record.seq}\\n\")\n",
    "        with open(f\"first/right_{file_base}\", 'a') as file:\n",
    "                file.write(f\">{record.id}\\n{right_subseq_record.seq}\\n\")\n",
    "\n",
    "def correct_first_part(fq=True, db=True, do_blast=True):\n",
    "    os.system(\"mkdir -p blast\")\n",
    "    os.system(\"mkdir -p first\")\n",
    "    if fq:\n",
    "        for file in ont:\n",
    "            os.system(f\"seqtk seq -a ont/{file}.fastq.gz > ont/{file}.fasta\")\n",
    "    if db:\n",
    "        print('okno')\n",
    "        for file in ont:\n",
    "            print(f\"makeblastdb -in ont/{file}.fasta -dbtype nucl -out ont/{file}\")\n",
    "            os.system(f\"makeblastdb -in ont/{file}.fasta -dbtype nucl -out ont/{file}\")\n",
    "    if do_blast:\n",
    "        print('okno2')\n",
    "        for file in ont:\n",
    "            print(\n",
    "                f\"blastn -num_threads 20 -outfmt 6 -query {te_path} -db ont/{file} \" \\\n",
    "                f\"-out blast/TE_{file}_{te_name}.bl -dust no -perc_identity 0.9\"\n",
    "            )\n",
    "            os.system(\n",
    "                f\"blastn -num_threads 20 -outfmt 6 -query {te_path} -db ont/{file} \" \\\n",
    "                f\"-out blast/TE_{file}_{te_name}.bl -dust no -perc_identity 0.9\"\n",
    "            )\n",
    "\n",
    "    for o in ont:\n",
    "        te_ont_bl = read_fasta(f\"blast/TE_{o}_{te_name}.bl\")\n",
    "        te_ont_bl = te_ont_bl[te_ont_bl.qseqid.str.contains(te_name)]\n",
    "        plus_te_ont_bl = te_ont_bl[te_ont_bl.sstart < te_ont_bl.send]\n",
    "        minus_te_ont_bl = te_ont_bl[te_ont_bl.sstart > te_ont_bl.send]\n",
    "    \n",
    "    \n",
    "        plus_te_ont_bed = plus_te_ont_bl[['sseqid', 'sstart', 'send', 'qseqid']].sort_values(by=['sseqid', 'sstart', 'send'])\n",
    "        minus_te_ont_bed = minus_te_ont_bl[['sseqid', 'send', 'sstart', 'qseqid']].sort_values(by=['sseqid', 'send', 'sstart'])\n",
    "    \n",
    "        save_fasta(minus_te_ont_bed, f'first/m_raw_first_blast_{o}_{te_name}.bed')\n",
    "        save_fasta(plus_te_ont_bed, f'first/p_raw_first_blast_{o}_{te_name}.bed')\n",
    "\n",
    "        os.system(f\"bedtools merge -i first/p_raw_first_blast_{o}_{te_name}.bed -d 1500 -c 4 -o collapse  > first/p_first_blast_{o}_{te_name}.bed.merged\")\n",
    "        os.system(f\"bedtools merge -i first/m_raw_first_blast_{o}_{te_name}.bed -d 1500 -c 4 -o collapse  > first/m_first_blast_{o}_{te_name}.bed.merged\")\n",
    "\n",
    "        print(\"first part\")\n",
    "    \n",
    "        plus_merged = pd.read_csv(\n",
    "            f\"first/p_first_blast_{o}_{te_name}.bed.merged\", sep=\"\\t\", header=None).rename(\n",
    "            {0:'sseqid', 1:'sstart', 2:'send', 3:'qseqid'}, axis=1\n",
    "        )\n",
    "        minus_merged = pd.read_csv(\n",
    "            f\"first/m_first_blast_{o}_{te_name}.bed.merged\", sep=\"\\t\", header=None).rename(\n",
    "            {0:'sseqid', 1:'sstart', 2:'send', 3:'qseqid'}, axis=1\n",
    "        )\n",
    "    \n",
    "        plus_filtered_bed = plus_merged[(plus_merged.send - plus_merged.sstart) > from_te]\n",
    "        minus_filtered_bed = minus_merged[(minus_merged.send - minus_merged.sstart) > from_te]\n",
    "\n",
    "        # todo in this place\n",
    "\n",
    "        save_fasta(pd.DataFrame({'id': plus_filtered_bed['sseqid'].unique()}), f'first/p_seq_list_{o}_{te_name}.txt')\n",
    "        save_fasta(pd.DataFrame({'id': minus_filtered_bed['sseqid'].unique()}), f'first/m_seq_list_{o}_{te_name}.txt')\n",
    "        #df.to_csv(filename, sep='\\t', index=False, encoding='utf-8', header=False)\n",
    "        print(\"second part\")\n",
    "    \n",
    "        #long\n",
    "        # it's filtered only our fasta\n",
    "        os.system(f\"seqtk subseq ont/{o}.fasta first/p_seq_list_{o}_{te_name}.txt > first/p_normal_{o}_{te_name}.fasta\")\n",
    "        os.system(f\"seqtk subseq ont/{o}.fasta first/m_seq_list_{o}_{te_name}.txt > first/m_normal_{o}_{te_name}.fasta\")\n",
    "        print(\"1 part\")\n",
    "\n",
    "        create_subsequent_fasta(f\"first/p_normal_{o}_{te_name}.fasta\", plus_filtered_bed,  f\"p_{o}_{te_name}.fasta\")\n",
    "        create_subsequent_fasta(f\"first/m_normal_{o}_{te_name}.fasta\", minus_filtered_bed,  f\"m_{o}_{te_name}.fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c1005ca-6ede-46f2-944b-305620683911",
   "metadata": {},
   "source": [
    "# Second part of analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "612515d0-bf15-42b3-a293-16526c8f9e6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\"\"\"\n",
    "x(start_df, end_df, strain, ont_file) #cale 2 df\n",
    "xx(start_record, end_df, strain) # jesli znajdzie dopasowuje prawo\n",
    "xxx(start_record, end_record) # True False\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "#second part\n",
    "class Blast:\n",
    "    @staticmethod\n",
    "    def make_db(fasta_file, out):\n",
    "        os.system(f\"makeblastdb -in {fasta_file} -dbtype nucl -out {out}\")\n",
    "\n",
    "    @staticmethod\n",
    "    def run(query, db, out):\n",
    "        os.system(f\"blastn -num_threads 20 -outfmt 6 -query {query} -db {db} -out {out} -perc_identity 0.9\")\n",
    "        return pd.read_csv(out, sep=\"\\t\", header=None).rename(headers_dict, axis=1)\n",
    "\n",
    "class SecondService:\n",
    "    def __init__(self, filename, te_name):\n",
    "        self.filename = filename\n",
    "        self.te_name = te_name\n",
    "\n",
    "        self.plus_te_start_records = {}\n",
    "        self.plus_te_end_records = {}\n",
    "        self.minus_te_start_records = {}\n",
    "        self.minus_te_end_records = {}\n",
    "\n",
    "    @property\n",
    "    def used_ids(self):\n",
    "        used_records = set()\n",
    "        for _, records in self.plus_te_start_records.items():\n",
    "            for record in records:\n",
    "                used_records.add(record.qseqid)\n",
    "        for _, records in self.minus_te_start_records.items():\n",
    "            for record in records:\n",
    "                used_records.add(record.qseqid)\n",
    "        for _, records in self.plus_te_end_records.items():\n",
    "            for record in records:\n",
    "                used_records.add(record.qseqid)\n",
    "        for _, records in self.minus_te_end_records.items():\n",
    "            for record in records:\n",
    "                used_records.add(record.qseqid)\n",
    "        return used_records\n",
    "    \n",
    "    def add_start(self, record, ont_file, is_plus_strain):\n",
    "        if is_plus_strain:\n",
    "            if self.plus_te_start_records.get(ont_file) is None:\n",
    "                self.plus_te_start_records[ont_file] = [record]\n",
    "            else:\n",
    "                self.plus_te_start_records[ont_file].append(record)\n",
    "        else:\n",
    "            if self.minus_te_start_records.get(ont_file) is None:\n",
    "                self.minus_te_start_records[ont_file] = [record]\n",
    "            else:\n",
    "                self.minus_te_start_records[ont_file].append(record)\n",
    "\n",
    "    def add_end(self, record, ont_file, is_plus_strain):\n",
    "        if is_plus_strain:\n",
    "            if self.plus_te_end_records.get(ont_file) is None:\n",
    "                self.plus_te_end_records[ont_file] = [record]\n",
    "            else:\n",
    "                self.plus_te_end_records[ont_file].append(record)\n",
    "        else:\n",
    "            if self.minus_te_end_records.get(ont_file) is None:\n",
    "                self.minus_te_end_records[ont_file] = [record]\n",
    "            else:\n",
    "                self.minus_te_end_records[ont_file].append(record)\n",
    "        \n",
    "    \n",
    "    def get_start_field(self, is_plus_strain):\n",
    "        return \"send\" if is_plus_strain else \"sstart\"\n",
    "\n",
    "    def get_end_field(self, is_plus_strain):\n",
    "        return \"sstart\" if is_plus_strain else \"send\"\n",
    "\n",
    "    def if_match(self, start_record, end_record, is_plus_strain):\n",
    "        start = getattr(start_record, self.get_start_field(is_plus_strain))\n",
    "        end = getattr(end_record, self.get_end_field(is_plus_strain))\n",
    "        return (end - start < to_te and end - start > from_te) or abs(end - start) < 200\n",
    "\n",
    "    def match_rec_df(self, start_record, end_df, is_plus_strain):\n",
    "        chromosome = start_record.sseqid\n",
    "        r_df =  end_df[(end_df.sseqid == chromosome)]\n",
    "        for right_row in r_df.itertuples():\n",
    "            if self.if_match(start_record, right_row, is_plus_strain):\n",
    "                return right_row\n",
    "        return None\n",
    "\n",
    "    '''\n",
    "    def match_not_ideal_pairs(self, left_df, right_df, is_plus_strain):\n",
    "        start_df = left_df.sort_values(by=['sseqid', 'length', self.get_start_field(is_plus_strain)])\n",
    "        end_df = right_df.sort_values(by=['sseqid', 'length', self.get_end_field(is_plus_strain)])\n",
    "        if not len(start_df) or not len(end_df):\n",
    "            return False\n",
    "            if not len(start_df):\n",
    "                if is_plus_strain:\n",
    "                    self.plus_te_end_records.append(end_df.iloc[0])\n",
    "                else:\n",
    "                    self.minus_te_end_records.append(end_df.iloc[0])\n",
    "                return True\n",
    "            elif not len(end_df):\n",
    "                if is_plus_strain:\n",
    "                    self.plus_te_start_records.append(start_df.iloc[0])\n",
    "                else:\n",
    "                    self.minus_te_start_records.append(start_df.iloc[0])\n",
    "                return True\n",
    "    \n",
    "            for chromosome in start_df.sseqid.unique():\n",
    "                l_df =  start_df[(start_df.sseqid == chromosome)]\n",
    "                r_df =  end_df[(end_df.sseqid == chromosome)]\n",
    "                \n",
    "        return False\n",
    "    '''\n",
    "    \n",
    "    def match_ideal_pairs(self, left_df, right_df, ont_file, is_plus_strain):\n",
    "        # add left and right if find\n",
    "        # iterate chromosome\n",
    "        # only one id\n",
    "        start_df = left_df.sort_values(by=['sseqid', 'length', self.get_start_field(is_plus_strain)])\n",
    "        end_df = right_df.sort_values(by=['sseqid', 'length', self.get_end_field(is_plus_strain)])\n",
    "        if not len(start_df) or not len(end_df):\n",
    "            return False\n",
    "        for chromosome in start_df.sseqid.unique():\n",
    "            l_df =  start_df[(start_df.sseqid == chromosome)]\n",
    "            r_df =  end_df[(end_df.sseqid == chromosome)]\n",
    "            for start_row in l_df.itertuples():\n",
    "                matched_right_record = self.match_rec_df(start_row, r_df, is_plus_strain)\n",
    "                if  matched_right_record is not None:\n",
    "                    if is_plus_strain:\n",
    "                        self.add_start(start_row, ont_file, is_plus_strain)\n",
    "                        self.add_end(matched_right_record, ont_file, is_plus_strain)\n",
    "                    return True\n",
    "                \n",
    "        return False\n",
    "    \n",
    "    def save_to_file(self, record, is_start, is_plus_strain, ont_file):\n",
    "        if record is not None:\n",
    "            r = {\n",
    "                \"chromosome\":record.sseqid,\n",
    "                \"strain\":\"plus\" if is_plus_strain else \"minus\",\n",
    "                \"start\":getattr(record, self.get_start_field(is_plus_strain)) if is_start else None,\n",
    "                \"stop\":getattr(record, self.get_end_field(is_plus_strain)) if not is_start else None,\n",
    "                \"ont_id\":f\"{ont_file}:{record.qseqid}\",\n",
    "            }\n",
    "            with open(self.filename, \"a\") as f_out:\n",
    "                f_out.write(\n",
    "                    f\"{r['chromosome']}\\t{r['strain']}\\t{r['start']}\\t{r['stop']}\\t{r['ont_id']}\\n\"\n",
    "                )\n",
    "    \n",
    "    def match_dfs(self, left_df, right_df, ont_file, is_plus_strain):\n",
    "        # iterate ont_ids\n",
    "        all_ids = set(left_df.qseqid.unique()).union(set(right_df.qseqid.unique()))\n",
    "        # msstch 1\n",
    "        for ont_id in all_ids:\n",
    "            left = left_df[left_df.qseqid == ont_id]\n",
    "            right = right_df[right_df.qseqid == ont_id]\n",
    "            self.match_ideal_pairs(left, right, ont_file, is_plus_strain)\n",
    "\n",
    "        \n",
    "        # todo - now\n",
    "        not_used_start_records = left_df[~left_df.qseqid.isin(self.used_ids)]\n",
    "        not_used_end_records = right_df[~right_df.qseqid.isin(self.used_ids)]\n",
    "        used_start_records = left_df[left_df.qseqid.isin(self.used_ids)]\n",
    "        used_end_records = right_df[right_df.qseqid.isin(self.used_ids)]\n",
    "        for not_used_start_record in not_used_start_records.itertuples():\n",
    "            for used_start_record in used_start_records.itertuples():\n",
    "                start1 = getattr(not_used_start_record, self.get_start_field(is_plus_strain))\n",
    "                start2 = getattr(used_start_record, self.get_start_field(is_plus_strain))\n",
    "                if abs(start1 - start2) < 300:\n",
    "                    print('konik')\n",
    "        \n",
    "        \n",
    "    \n",
    "    # main function\n",
    "    def xxx(self, left_blast, right_blast, ont_file):\n",
    "        left_blast =  left_blast[left_blast.length > 600]\n",
    "        right_blast =  right_blast[right_blast.length > 600]\n",
    "    \n",
    "        plus_left = left_blast[left_blast.sstart <= left_blast.send]\n",
    "        plus_left[\"strain\"] = \"+\"\n",
    "        plus_right =  right_blast[right_blast.sstart <= right_blast.send]\n",
    "        plus_right[\"strain\"] = \"+\"\n",
    "        \n",
    "        minus_left = left_blast[left_blast.sstart > left_blast.send]\n",
    "        minus_left[\"strain\"] = \"-\"\n",
    "        minus_right = right_blast[right_blast.sstart > right_blast.send]\n",
    "        minus_right[\"strain\"] = \"-\"\n",
    "\n",
    "        self.match_dfs(plus_left, plus_right, ont_file, True)\n",
    "        self.match_dfs(minus_right, minus_left, ont_file, False)\n",
    "\n",
    "    def ont_to_genome_blast(self, fasta_filename, blast_filename):\n",
    "        return Blast.run(\n",
    "            fasta_filename,\n",
    "            f\"genome/masked_genome\",\n",
    "            blast_filename,\n",
    "        ).sort_values(['qseqid', 'bitscore'], ascending=False).groupby(['qseqid'], as_index=False).first()\n",
    "\n",
    "    def blast(self, o, blast_db=True):\n",
    "        left_m = self.ont_to_genome_blast(\n",
    "            f\"first/left_m_{o}_{self.te_name}.fasta\",\n",
    "            f\"second/left_m_{o}_{self.te_name}_genome.bl\",\n",
    "        )\n",
    "        right_m = self.ont_to_genome_blast(\n",
    "            f\"first/right_m_{o}_{self.te_name}.fasta\",\n",
    "            f\"second/right_m_{o}_{self.te_name}_genome.bl\",\n",
    "        )\n",
    "        left_p = self.ont_to_genome_blast(\n",
    "            f\"first/left_p_{o}_{self.te_name}.fasta\",\n",
    "            f\"second/left_p_{o}_{self.te_name}_genome.bl\",\n",
    "        )\n",
    "        right_p = self.ont_to_genome_blast(\n",
    "            f\"first/right_p_{o}_{self.te_name}.fasta\",\n",
    "            f\"second/right_p_{o}_{self.te_name}_genome.bl\",\n",
    "        )\n",
    "        return [\n",
    "            [left_m, right_m], [left_p, right_p]\n",
    "        ]\n",
    "    \n",
    "    def run(self, blast_db=True):\n",
    "\n",
    "        raw_report_header = [\"chromosome\", \"strain\", \"start\", \"stop\", \"ont_id\"]\n",
    "        header_to_save = '\\t'.join(raw_report_header)\n",
    "        with open(self.filename, \"w\") as f_out:\n",
    "            f_out.write(\n",
    "                f\"{header_to_save}\\n\"\n",
    "            )\n",
    "    \n",
    "        report = {}\n",
    "        os.system(\"mkdir -p final\")\n",
    "        os.system(\"mkdir -p second\")\n",
    "        pd.options.mode.copy_on_write = True\n",
    "        if blast_db:\n",
    "            Blast.make_db(genome_path, \"genome/masked_genome\")\n",
    "        for o in ont:\n",
    "            for left_blast, right_blast in self.blast(o, blast_db):\n",
    "                self.xxx(left_blast, right_blast, o)\n",
    "    \n",
    "        for ont_file, start_records in self.plus_te_start_records.items():\n",
    "            for start_record in start_records:\n",
    "                self.save_to_file(start_record, True, True, ont_file)\n",
    "        for ont_file, end_records in self.plus_te_end_records.items():\n",
    "            for end_record in end_records:\n",
    "                self.save_to_file(end_record, False, True, ont_file)\n",
    "            \n",
    "        for ont_file, start_records in self.minus_te_start_records.items():\n",
    "            for start_record in start_records:\n",
    "                self.save_to_file(start_record, True, False, ont_file)\n",
    "        for ont_file, end_records in self.minus_te_end_records.items():\n",
    "            for end_record in end_records:\n",
    "                self.save_to_file(end_record, False, False, ont_file)\n",
    "        #report.to_csv(self.filename, sep='\\t', index=False, encoding='utf-8', header=True)\n",
    "        return pd.read_csv(self.filename, sep=\"\\t\")\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "e46200c2-c003-4386-8379-ac35b98b5945",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n",
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n",
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n",
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n",
      "Warning: [blastn] Query_28 a4e33dfa-5ddb-4e.. : Sequence contains no data \n",
      "Warning: [blastn] Query_68 a4e33dfa-5ddb-4e.. : Sequence contains no data \n",
      "Warning: [blastn] Query_108 a4e33dfa-5ddb-4.. : Sequence contains no data \n",
      "Warning: [blastn] Query_148 a4e33dfa-5ddb-4.. : Sequence contains no data \n",
      "Warning: [blastn] Query_188 a4e33dfa-5ddb-4.. : Sequence contains no data \n",
      "Warning: [blastn] Query_228 a4e33dfa-5ddb-4.. : Sequence contains no data \n",
      "Warning: [blastn] Query_268 a4e33dfa-5ddb-4.. : Sequence contains no data \n",
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n",
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n",
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n",
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "konik\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n",
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n",
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n",
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n",
      "Warning: [blastn] Query_21 9cd45113-0200-4e.. : Sequence contains no data \n",
      "Warning: [blastn] Query_50 9cd45113-0200-4e.. : Sequence contains no data \n",
      "Warning: [blastn] Query_79 9cd45113-0200-4e.. : Sequence contains no data \n",
      "Warning: [blastn] Query_108 9cd45113-0200-4.. : Sequence contains no data \n",
      "Warning: [blastn] Query_137 9cd45113-0200-4.. : Sequence contains no data \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "konik\n",
      "konik\n",
      "konik\n",
      "konik\n",
      "konik\n",
      "konik\n",
      "konik\n",
      "konik\n",
      "konik\n",
      "konik\n",
      "konik\n",
      "konik\n"
     ]
    }
   ],
   "source": [
    "filename = \"zizwa\"\n",
    "service = SecondService(filename, te_name)\n",
    "raw_report = service.run(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "c2e7e57d-fa78-4461-9b14-9cb7b915da0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "90"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(raw_report)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "524e3122-2ac4-4d91-a7f1-514e1d835848",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strain</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>ont_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome9</td>\n",
       "      <td>plus</td>\n",
       "      <td>28530068.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K10B:e9b31c79-83c5-4cbf-a699-684507b19baa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chromosome9</td>\n",
       "      <td>plus</td>\n",
       "      <td>28529338.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K10B:10e616cd-c2ca-42e2-a391-8078ac0d8956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chromosome9</td>\n",
       "      <td>plus</td>\n",
       "      <td>9220466.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K10B:070100af-93c2-4c1c-b426-0e6852ff5740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chromosome9</td>\n",
       "      <td>plus</td>\n",
       "      <td>28529396.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K10B:c20f1362-fbcb-4230-8a32-73659bc06318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chromosome5</td>\n",
       "      <td>plus</td>\n",
       "      <td>41745603.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K10B:d69243ea-29eb-45f5-9a67-cb1adf455e12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chromosome strain       start  stop  \\\n",
       "0  chromosome9   plus  28530068.0   NaN   \n",
       "1  chromosome9   plus  28529338.0   NaN   \n",
       "2  chromosome9   plus   9220466.0   NaN   \n",
       "3  chromosome9   plus  28529396.0   NaN   \n",
       "4  chromosome5   plus  41745603.0   NaN   \n",
       "\n",
       "                                      ont_id  \n",
       "0  K10B:e9b31c79-83c5-4cbf-a699-684507b19baa  \n",
       "1  K10B:10e616cd-c2ca-42e2-a391-8078ac0d8956  \n",
       "2  K10B:070100af-93c2-4c1c-b426-0e6852ff5740  \n",
       "3  K10B:c20f1362-fbcb-4230-8a32-73659bc06318  \n",
       "4  K10B:d69243ea-29eb-45f5-9a67-cb1adf455e12  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_report.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "daadbc3a-2b84-48bb-ad0d-450656bb3bb4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "\n",
    "def get_first_group(df, field):\n",
    "    first_row = df.iloc[0]\n",
    "    field_value = first_row[field]\n",
    "    #group = df[(df[field] > field_value - 1700) & (df[field] < field_value + 1700)]\n",
    "    #new_df = df[(df[field] <= field_value - 1700) | (df[field] >= field_value + 1700)]\n",
    "    group = df[(df[field] > field_value - 1000) & (df[field] < field_value + 1000)]\n",
    "    new_df = df[(df[field] <= field_value - 1000) | (df[field] >= field_value + 1000)]\n",
    "    return new_df, group\n",
    "\n",
    "def get_groups_dict(groups, field):\n",
    "    group_dict = {}\n",
    "    for group in groups:\n",
    "        first_row = group.iloc[0]\n",
    "        group_dict[first_row.chromosome, int(group[field].mode().median())] = group\n",
    "    return group_dict\n",
    "\n",
    "def get_groups(records, field):\n",
    "    all_records = records[raw_report[field].notnull()]\n",
    "    groups = []\n",
    "    for chromosome in all_records.chromosome.unique():\n",
    "        remaining = all_records[all_records.chromosome == chromosome]\n",
    "        while len(remaining):\n",
    "            g = get_first_group(remaining, field)\n",
    "            remaining = g[0]\n",
    "            groups.append(g[1])\n",
    "    return groups\n",
    "\n",
    "\n",
    "\n",
    "def if_match(start_chromosome, end_chromosome, start, end):\n",
    "    if start_chromosome == end_chromosome:\n",
    "        if (end - start > from_te and end - start < to_te) or (abs(end - start) < 200):\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "def create_record(chromosome, start, end, start_records, end_records):\n",
    "    start_ids = set([ont_id for ont_id in start_records.ont_id])\n",
    "    end_ids = set([ont_id for ont_id in end_records.ont_id])\n",
    "    if start_ids.intersection(end_ids):\n",
    "        pkt = 1\n",
    "    elif set([x.split(\":\")[0] for x in start_ids]).intersection([x.split(\":\")[0] for x in end_ids]):\n",
    "        pkt = 2\n",
    "    elif start_ids and start_ids:\n",
    "        pkt = 3\n",
    "    else:\n",
    "        pkt = 4\n",
    "    return  {\n",
    "        \"pkt\":pkt,\n",
    "        \"chromosome\": chromosome,\n",
    "        \"start\":start,\n",
    "        \"stop\":end,\n",
    "        \"length\":end-start,\n",
    "        \"left_n\":len(start_records),\n",
    "        \"left_rec\":\",\".join([ont_id for ont_id in start_records.ont_id]),\n",
    "        \"right_n\":len(end_records),\n",
    "        \"right_rec\":\",\".join([ont_id for ont_id in end_records.ont_id]),\n",
    "        }\n",
    "\n",
    "\n",
    "def match(start_dict, end_dict):\n",
    "    report_list = []\n",
    "    start_remaining = start_dict.copy()\n",
    "    end_remaining = end_dict.copy()\n",
    "    for (start_chromosome, start), start_records in start_dict.items():\n",
    "        end_dict = end_remaining.copy()\n",
    "        for (end_chromosome, end), end_records in end_dict.items():\n",
    "            if if_match(start_chromosome, end_chromosome, start, end):\n",
    "                report_list.append(\n",
    "                    create_record(\n",
    "                        start_chromosome,\n",
    "                        start,\n",
    "                        end,\n",
    "                        start_records,\n",
    "                        end_records\n",
    "                    )\n",
    "                )\n",
    "                del start_remaining[start_chromosome, start]\n",
    "                del end_remaining[end_chromosome, end]\n",
    "                continue\n",
    "        #todo if not match?\n",
    "    return report_list\n",
    "\n",
    "def save_report(report_list, filename):\n",
    "    report_header = [\"pkt\", \"chromosome\", \"start\", \"stop\", \"length\",  \"left_n\", \"left_rec\", \"right_n\", \"right_rec\"]\n",
    "    header_to_save = '\\t'.join(report_header)\n",
    "    with open(filename, \"w\") as f_out:\n",
    "        f_out.write(\n",
    "            f\"{header_to_save}\\n\"\n",
    "        )\n",
    "    for r in report_list:\n",
    "        with open(filename2, \"a\") as f_out:\n",
    "            f_out.write(\n",
    "                f\"{r['pkt']}\\t{r['chromosome']}\\t{r['start']}\\t{r['stop']}\\t{r['length']}\\t\"\n",
    "                f\"{r['left_n']}\\t{r['left_rec']}\\t{r['right_n']}\\t{r['right_rec']}\\n\"\n",
    "            )\n",
    "\n",
    "def save_first_point_report(filename):\n",
    "    start_groups = get_groups(raw_report, \"start\")\n",
    "    end_groups = get_groups(raw_report, \"stop\")\n",
    "\n",
    "    start_dict = get_groups_dict(start_groups, \"start\")\n",
    "    end_dict = get_groups_dict(end_groups, \"stop\")\n",
    "\n",
    "    report_list = match(start_dict, end_dict)\n",
    "\n",
    "    \n",
    "    save_report(report_list, filename)\n",
    "\n",
    "    report = pd.read_csv(filename, sep=\"\\t\").sort_values(by=['chromosome', 'pkt', 'start'])\n",
    "    report.to_csv(filename, sep='\\t', index=False, encoding='utf-8', header=True)\n",
    "    return report\n",
    "\n",
    "filename2 = \"moli2\"\n",
    "first_report = save_first_point_report(filename2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "d11cfd59-5f78-4c76-8a9e-97c2c77bcbfe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pkt</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>length</th>\n",
       "      <th>left_n</th>\n",
       "      <th>left_rec</th>\n",
       "      <th>right_n</th>\n",
       "      <th>right_rec</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome1</td>\n",
       "      <td>1895035</td>\n",
       "      <td>1895057</td>\n",
       "      <td>22</td>\n",
       "      <td>3</td>\n",
       "      <td>LK5_1_3:ad199210-fea3-4427-b40a-288d4cb970a8,L...</td>\n",
       "      <td>3</td>\n",
       "      <td>LK5_1_3:ad199210-fea3-4427-b40a-288d4cb970a8,L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome1</td>\n",
       "      <td>18795335</td>\n",
       "      <td>18799060</td>\n",
       "      <td>3725</td>\n",
       "      <td>2</td>\n",
       "      <td>LK5_1_3:f9d75fb2-63a6-4a2d-b898-71b4c2099e3d,L...</td>\n",
       "      <td>2</td>\n",
       "      <td>LK5_1_3:f9d75fb2-63a6-4a2d-b898-71b4c2099e3d,L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome2</td>\n",
       "      <td>24991939</td>\n",
       "      <td>24991944</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:aa99dd0d-ba11-4f47-a79e-5dfedc9da5e2</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:aa99dd0d-ba11-4f47-a79e-5dfedc9da5e2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome2</td>\n",
       "      <td>30539628</td>\n",
       "      <td>30544358</td>\n",
       "      <td>4730</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:b647706c-82df-4164-9168-4b200501ca6d</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:b647706c-82df-4164-9168-4b200501ca6d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome3</td>\n",
       "      <td>27718180</td>\n",
       "      <td>27722607</td>\n",
       "      <td>4427</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:6c07ed7c-c720-49ac-9e4b-b83309306bae</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:6c07ed7c-c720-49ac-9e4b-b83309306bae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome3</td>\n",
       "      <td>35158725</td>\n",
       "      <td>35158723</td>\n",
       "      <td>-2</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:cf60c0dc-a60a-4a67-acb7-28845f6eeb4c</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:cf60c0dc-a60a-4a67-acb7-28845f6eeb4c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome3</td>\n",
       "      <td>41159456</td>\n",
       "      <td>41164176</td>\n",
       "      <td>4720</td>\n",
       "      <td>3</td>\n",
       "      <td>K10B:7e4c3981-29d5-40a6-aad0-59e1e9322d19,K10B...</td>\n",
       "      <td>3</td>\n",
       "      <td>K10B:7e4c3981-29d5-40a6-aad0-59e1e9322d19,K10B...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome4</td>\n",
       "      <td>4759534</td>\n",
       "      <td>4759532</td>\n",
       "      <td>-2</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:941656fa-26ed-4578-8aab-59824b70f133</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:941656fa-26ed-4578-8aab-59824b70f133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome4</td>\n",
       "      <td>35867416</td>\n",
       "      <td>35867451</td>\n",
       "      <td>35</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:a7a007a4-c486-4df5-854d-69f74d61d3d1</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:a7a007a4-c486-4df5-854d-69f74d61d3d1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome5</td>\n",
       "      <td>8967214</td>\n",
       "      <td>8971940</td>\n",
       "      <td>4726</td>\n",
       "      <td>5</td>\n",
       "      <td>LK5_1_3:02a6bc39-ecc1-41f0-b5ca-c26c015ff4b1,L...</td>\n",
       "      <td>5</td>\n",
       "      <td>LK5_1_3:02a6bc39-ecc1-41f0-b5ca-c26c015ff4b1,L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome5</td>\n",
       "      <td>34862567</td>\n",
       "      <td>34862563</td>\n",
       "      <td>-4</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:dc8aa270-492a-4c08-933a-5c3b8eaa980e</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:dc8aa270-492a-4c08-933a-5c3b8eaa980e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome5</td>\n",
       "      <td>41745603</td>\n",
       "      <td>41745600</td>\n",
       "      <td>-3</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:d69243ea-29eb-45f5-9a67-cb1adf455e12</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:d69243ea-29eb-45f5-9a67-cb1adf455e12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome6</td>\n",
       "      <td>6237082</td>\n",
       "      <td>6237101</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>K10f:a4c02692-8649-450c-87ec-9d8d532f47cf</td>\n",
       "      <td>1</td>\n",
       "      <td>K10f:a4c02692-8649-450c-87ec-9d8d532f47cf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome7</td>\n",
       "      <td>12013941</td>\n",
       "      <td>12013940</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:92e0de8b-6882-4283-87a2-491864754ad3</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:92e0de8b-6882-4283-87a2-491864754ad3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome7</td>\n",
       "      <td>13174817</td>\n",
       "      <td>13174838</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:29af0fb0-6c1a-44d8-8425-57c5f585641d</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:29af0fb0-6c1a-44d8-8425-57c5f585641d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome8</td>\n",
       "      <td>11878585</td>\n",
       "      <td>11878581</td>\n",
       "      <td>-4</td>\n",
       "      <td>1</td>\n",
       "      <td>K10f:114221bc-ac48-4940-8416-0ba3ab107414</td>\n",
       "      <td>1</td>\n",
       "      <td>K10f:114221bc-ac48-4940-8416-0ba3ab107414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome8</td>\n",
       "      <td>22629055</td>\n",
       "      <td>22629051</td>\n",
       "      <td>-4</td>\n",
       "      <td>1</td>\n",
       "      <td>K10f:75a9a3f3-13e0-4403-a2bb-a1e6763fdadf</td>\n",
       "      <td>1</td>\n",
       "      <td>K10f:75a9a3f3-13e0-4403-a2bb-a1e6763fdadf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome8</td>\n",
       "      <td>31372224</td>\n",
       "      <td>31372250</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:ac77f568-5294-480f-8ed6-410cdc82fdde</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:ac77f568-5294-480f-8ed6-410cdc82fdde</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome9</td>\n",
       "      <td>9220466</td>\n",
       "      <td>9226373</td>\n",
       "      <td>5907</td>\n",
       "      <td>9</td>\n",
       "      <td>K10B:070100af-93c2-4c1c-b426-0e6852ff5740,K10B...</td>\n",
       "      <td>9</td>\n",
       "      <td>K10B:070100af-93c2-4c1c-b426-0e6852ff5740,K10B...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome9</td>\n",
       "      <td>28529370</td>\n",
       "      <td>28534569</td>\n",
       "      <td>5199</td>\n",
       "      <td>8</td>\n",
       "      <td>K10B:e9b31c79-83c5-4cbf-a699-684507b19baa,K10B...</td>\n",
       "      <td>8</td>\n",
       "      <td>K10B:e9b31c79-83c5-4cbf-a699-684507b19baa,K10B...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome9</td>\n",
       "      <td>31434581</td>\n",
       "      <td>31434577</td>\n",
       "      <td>-4</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:66e7bd3c-b6e7-476e-98f7-79d006377107</td>\n",
       "      <td>1</td>\n",
       "      <td>K10B:66e7bd3c-b6e7-476e-98f7-79d006377107</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    pkt   chromosome     start      stop  length  left_n  \\\n",
       "19    1  chromosome1   1895035   1895057      22       3   \n",
       "20    1  chromosome1  18795335  18799060    3725       2   \n",
       "7     1  chromosome2  24991939  24991944       5       1   \n",
       "6     1  chromosome2  30539628  30544358    4730       1   \n",
       "8     1  chromosome3  27718180  27722607    4427       1   \n",
       "9     1  chromosome3  35158725  35158723      -2       1   \n",
       "10    1  chromosome3  41159456  41164176    4720       3   \n",
       "16    1  chromosome4   4759534   4759532      -2       1   \n",
       "17    1  chromosome4  35867416  35867451      35       1   \n",
       "5     1  chromosome5   8967214   8971940    4726       5   \n",
       "4     1  chromosome5  34862567  34862563      -4       1   \n",
       "3     1  chromosome5  41745603  41745600      -3       1   \n",
       "18    1  chromosome6   6237082   6237101      19       1   \n",
       "15    1  chromosome7  12013941  12013940      -1       1   \n",
       "14    1  chromosome7  13174817  13174838      21       1   \n",
       "12    1  chromosome8  11878585  11878581      -4       1   \n",
       "13    1  chromosome8  22629055  22629051      -4       1   \n",
       "11    1  chromosome8  31372224  31372250      26       1   \n",
       "1     1  chromosome9   9220466   9226373    5907       9   \n",
       "0     1  chromosome9  28529370  28534569    5199       8   \n",
       "2     1  chromosome9  31434581  31434577      -4       1   \n",
       "\n",
       "                                             left_rec  right_n  \\\n",
       "19  LK5_1_3:ad199210-fea3-4427-b40a-288d4cb970a8,L...        3   \n",
       "20  LK5_1_3:f9d75fb2-63a6-4a2d-b898-71b4c2099e3d,L...        2   \n",
       "7           K10B:aa99dd0d-ba11-4f47-a79e-5dfedc9da5e2        1   \n",
       "6           K10B:b647706c-82df-4164-9168-4b200501ca6d        1   \n",
       "8           K10B:6c07ed7c-c720-49ac-9e4b-b83309306bae        1   \n",
       "9           K10B:cf60c0dc-a60a-4a67-acb7-28845f6eeb4c        1   \n",
       "10  K10B:7e4c3981-29d5-40a6-aad0-59e1e9322d19,K10B...        3   \n",
       "16          K10B:941656fa-26ed-4578-8aab-59824b70f133        1   \n",
       "17          K10B:a7a007a4-c486-4df5-854d-69f74d61d3d1        1   \n",
       "5   LK5_1_3:02a6bc39-ecc1-41f0-b5ca-c26c015ff4b1,L...        5   \n",
       "4           K10B:dc8aa270-492a-4c08-933a-5c3b8eaa980e        1   \n",
       "3           K10B:d69243ea-29eb-45f5-9a67-cb1adf455e12        1   \n",
       "18          K10f:a4c02692-8649-450c-87ec-9d8d532f47cf        1   \n",
       "15          K10B:92e0de8b-6882-4283-87a2-491864754ad3        1   \n",
       "14          K10B:29af0fb0-6c1a-44d8-8425-57c5f585641d        1   \n",
       "12          K10f:114221bc-ac48-4940-8416-0ba3ab107414        1   \n",
       "13          K10f:75a9a3f3-13e0-4403-a2bb-a1e6763fdadf        1   \n",
       "11          K10B:ac77f568-5294-480f-8ed6-410cdc82fdde        1   \n",
       "1   K10B:070100af-93c2-4c1c-b426-0e6852ff5740,K10B...        9   \n",
       "0   K10B:e9b31c79-83c5-4cbf-a699-684507b19baa,K10B...        8   \n",
       "2           K10B:66e7bd3c-b6e7-476e-98f7-79d006377107        1   \n",
       "\n",
       "                                            right_rec  \n",
       "19  LK5_1_3:ad199210-fea3-4427-b40a-288d4cb970a8,L...  \n",
       "20  LK5_1_3:f9d75fb2-63a6-4a2d-b898-71b4c2099e3d,L...  \n",
       "7           K10B:aa99dd0d-ba11-4f47-a79e-5dfedc9da5e2  \n",
       "6           K10B:b647706c-82df-4164-9168-4b200501ca6d  \n",
       "8           K10B:6c07ed7c-c720-49ac-9e4b-b83309306bae  \n",
       "9           K10B:cf60c0dc-a60a-4a67-acb7-28845f6eeb4c  \n",
       "10  K10B:7e4c3981-29d5-40a6-aad0-59e1e9322d19,K10B...  \n",
       "16          K10B:941656fa-26ed-4578-8aab-59824b70f133  \n",
       "17          K10B:a7a007a4-c486-4df5-854d-69f74d61d3d1  \n",
       "5   LK5_1_3:02a6bc39-ecc1-41f0-b5ca-c26c015ff4b1,L...  \n",
       "4           K10B:dc8aa270-492a-4c08-933a-5c3b8eaa980e  \n",
       "3           K10B:d69243ea-29eb-45f5-9a67-cb1adf455e12  \n",
       "18          K10f:a4c02692-8649-450c-87ec-9d8d532f47cf  \n",
       "15          K10B:92e0de8b-6882-4283-87a2-491864754ad3  \n",
       "14          K10B:29af0fb0-6c1a-44d8-8425-57c5f585641d  \n",
       "12          K10f:114221bc-ac48-4940-8416-0ba3ab107414  \n",
       "13          K10f:75a9a3f3-13e0-4403-a2bb-a1e6763fdadf  \n",
       "11          K10B:ac77f568-5294-480f-8ed6-410cdc82fdde  \n",
       "1   K10B:070100af-93c2-4c1c-b426-0e6852ff5740,K10B...  \n",
       "0   K10B:e9b31c79-83c5-4cbf-a699-684507b19baa,K10B...  \n",
       "2           K10B:66e7bd3c-b6e7-476e-98f7-79d006377107  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "first_report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 553,
   "id": "e8410a75-6069-46cf-8cd2-61f3c1148b57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chromosome2</td>\n",
       "      <td>30539632</td>\n",
       "      <td>30544357</td>\n",
       "      <td>4725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chromosome3</td>\n",
       "      <td>27717880</td>\n",
       "      <td>27722606</td>\n",
       "      <td>4726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chromosome5</td>\n",
       "      <td>8967214</td>\n",
       "      <td>8971939</td>\n",
       "      <td>4725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chromosome9</td>\n",
       "      <td>9220466</td>\n",
       "      <td>9226148</td>\n",
       "      <td>5682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chromosome9</td>\n",
       "      <td>28529339</td>\n",
       "      <td>28534568</td>\n",
       "      <td>5229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome1</td>\n",
       "      <td>18795334</td>\n",
       "      <td>18799058</td>\n",
       "      <td>3724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chromosome3</td>\n",
       "      <td>41159456</td>\n",
       "      <td>41164182</td>\n",
       "      <td>4726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chromosome4</td>\n",
       "      <td>35827649</td>\n",
       "      <td>35832374</td>\n",
       "      <td>4725</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0         1         2  length\n",
       "1   chromosome2  30539632  30544357    4725\n",
       "4   chromosome3  27717880  27722606    4726\n",
       "6   chromosome5   8967214   8971939    4725\n",
       "9   chromosome9   9220466   9226148    5682\n",
       "10  chromosome9  28529339  28534568    5229\n",
       "0   chromosome1  18795334  18799058    3724\n",
       "2   chromosome3  41159456  41164182    4726\n",
       "5   chromosome4  35827649  35832374    4725"
      ]
     },
     "execution_count": 553,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "masked_genome\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "59a2eff5-6cca-4787-a5df-7f7fcd43f203",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['pkt', 'chromosome', 'start', 'stop', 'length', 'is_ref', 'left_K10B_num', 'left_K10B_rec', 'right_K10B_num', 'right_K10B_rec', 'left_K10f_num', 'left_K10f_rec', 'right_K10f_num', 'right_K10f_rec', 'left_LK5_1_3_num', 'left_LK5_1_3_rec', 'right_LK5_1_3_num', 'right_LK5_1_3_rec']\n"
     ]
    }
   ],
   "source": [
    "def _is_ref(df, chromosome, start, stop):\n",
    "    return (\n",
    "        (df.chromosome == chromosome)\n",
    "        & (df.start < start + 500) & (df.start > start - 500)\n",
    "        & (df.stop < stop + 500) & (df.stop > stop - 500)\n",
    "    )\n",
    "\n",
    "def is_ref(df, reference):\n",
    "    is_reference = None\n",
    "    for rec in  masked_genome.itertuples():\n",
    "        chromosome = rec[1]\n",
    "        start = rec[2]\n",
    "        stop = rec[3]\n",
    "        part_is_reference = _is_ref(df, chromosome, start, stop)\n",
    "        is_reference = is_reference | part_is_reference if is_reference is not None else part_is_reference\n",
    "    return is_reference\n",
    "\n",
    "def get_rec(x, o):\n",
    "    records = []\n",
    "    for r in x.split(\",\"):\n",
    "        rec = r.split(\":\")\n",
    "        o_file = rec[0]\n",
    "        o_id = rec[1]\n",
    "        if o_file == o:\n",
    "            records.append(o_id)\n",
    "    return \",\".join(records)\n",
    "\n",
    "def get_num(x, o):\n",
    "    records = []\n",
    "    for r in x.split(\",\"):\n",
    "        rec = r.split(\":\")\n",
    "        o_file = rec[0]\n",
    "        o_id = rec[1]\n",
    "        if o_file == o:\n",
    "            records.append(o_id)\n",
    "    return len(records)\n",
    "\n",
    "def generate_second_report(filename_in, filename_out, reference):\n",
    "    report_header = [\"pkt\", \"chromosome\", \"start\", \"stop\", \"length\", \"is_ref\"]\n",
    "    df =  pd.read_csv(filename2, sep=\"\\t\")\n",
    "    for o in ont:\n",
    "        df[f\"left_{o}_num\"] = df.left_rec.apply(\n",
    "            get_num,\n",
    "            args = [o]\n",
    "        )\n",
    "        df[f\"left_{o}_rec\"] = df.left_rec.apply(\n",
    "            get_rec,\n",
    "            args = [o]\n",
    "        )\n",
    "        report_header.append(f\"left_{o}_num\")\n",
    "        report_header.append(f\"left_{o}_rec\")\n",
    "\n",
    "        df[f\"right_{o}_num\"] = df.right_rec.apply(\n",
    "            get_num,\n",
    "            args = [o]\n",
    "        )\n",
    "        df[f\"right_{o}_rec\"] = df.right_rec.apply(\n",
    "            get_rec,\n",
    "            args = [o]\n",
    "        )\n",
    "        report_header.append(f\"right_{o}_num\")\n",
    "        report_header.append(f\"right_{o}_rec\")\n",
    "\n",
    "    is_reff = is_ref(df, masked_genome)\n",
    "    df['is_ref'] = \"nonref\"\n",
    "    df.loc[(df.length > 600), 'is_ref'] = \"unknown\"    \n",
    "    df.loc[is_reff, 'is_ref'] = \"ref\"    \n",
    "\n",
    "    print(report_header)\n",
    "    report = df[report_header].sort_values(by=['chromosome', 'pkt', 'start'])\n",
    "    report.to_csv(filename_out, sep='\\t', index=False, encoding='utf-8', header=True)\n",
    "    return report\n",
    "        \n",
    "\n",
    "    \n",
    "filename3 = \"wuxi3\"\n",
    "xxx = generate_second_report(filename2, filename3, masked_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 469,
   "id": "694f2188-a780-4867-97d8-27b582be3df9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pkt</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>length</th>\n",
       "      <th>is_ref</th>\n",
       "      <th>left_K10B_num</th>\n",
       "      <th>left_K10B_rec</th>\n",
       "      <th>left_K10f_num</th>\n",
       "      <th>left_K10f_rec</th>\n",
       "      <th>left_LK5_1_3_num</th>\n",
       "      <th>left_LK5_1_3_rec</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome1</td>\n",
       "      <td>1895035</td>\n",
       "      <td>1895052</td>\n",
       "      <td>17</td>\n",
       "      <td>nonref</td>\n",
       "      <td>1</td>\n",
       "      <td>2c40afb0-7c90-4ddf-847c-a587e291d4a9</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>6</td>\n",
       "      <td>d7669e06-6916-435d-9925-6c207b9258ee,fe42f608-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome1</td>\n",
       "      <td>18795334</td>\n",
       "      <td>18799060</td>\n",
       "      <td>3726</td>\n",
       "      <td>ref</td>\n",
       "      <td>1</td>\n",
       "      <td>ca50976c-e581-4f3d-a225-d6fb73c2b031</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>5</td>\n",
       "      <td>c81615f3-be5b-4c2b-9b60-bad3cfe7feb4,a597c928-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome1</td>\n",
       "      <td>27115102</td>\n",
       "      <td>27115104</td>\n",
       "      <td>2</td>\n",
       "      <td>nonref</td>\n",
       "      <td>1</td>\n",
       "      <td>9ebca745-06aa-4187-9303-3fc73853bf74</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome1</td>\n",
       "      <td>40590596</td>\n",
       "      <td>40590593</td>\n",
       "      <td>-3</td>\n",
       "      <td>nonref</td>\n",
       "      <td>1</td>\n",
       "      <td>98c0c670-f831-449a-a2d7-6b50aad7d341</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome1</td>\n",
       "      <td>43955546</td>\n",
       "      <td>43955542</td>\n",
       "      <td>-4</td>\n",
       "      <td>nonref</td>\n",
       "      <td>1</td>\n",
       "      <td>bed145d3-72e9-4002-ba69-3a2a28356019</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pkt   chromosome     start      stop  length  is_ref  left_K10B_num  \\\n",
       "0    1  chromosome1   1895035   1895052      17  nonref              1   \n",
       "1    1  chromosome1  18795334  18799060    3726     ref              1   \n",
       "2    1  chromosome1  27115102  27115104       2  nonref              1   \n",
       "3    1  chromosome1  40590596  40590593      -3  nonref              1   \n",
       "4    1  chromosome1  43955546  43955542      -4  nonref              1   \n",
       "\n",
       "                          left_K10B_rec  left_K10f_num left_K10f_rec  \\\n",
       "0  2c40afb0-7c90-4ddf-847c-a587e291d4a9              0                 \n",
       "1  ca50976c-e581-4f3d-a225-d6fb73c2b031              0                 \n",
       "2  9ebca745-06aa-4187-9303-3fc73853bf74              0                 \n",
       "3  98c0c670-f831-449a-a2d7-6b50aad7d341              0                 \n",
       "4  bed145d3-72e9-4002-ba69-3a2a28356019              0                 \n",
       "\n",
       "   left_LK5_1_3_num                                   left_LK5_1_3_rec  \n",
       "0                 6  d7669e06-6916-435d-9925-6c207b9258ee,fe42f608-...  \n",
       "1                 5  c81615f3-be5b-4c2b-9b60-bad3cfe7feb4,a597c928-...  \n",
       "2                 0                                                     \n",
       "3                 0                                                     \n",
       "4                 0                                                     "
      ]
     },
     "execution_count": 469,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xxx.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 409,
   "id": "94e012fa-0eb6-4ac3-90b7-ebedc1074c2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pkt</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>length</th>\n",
       "      <th>is_ref</th>\n",
       "      <th>left_K10B_num</th>\n",
       "      <th>left_K10B_rec</th>\n",
       "      <th>left_K10f_num</th>\n",
       "      <th>left_K10f_rec</th>\n",
       "      <th>left_LK5_1_3_num</th>\n",
       "      <th>left_LK5_1_3_rec</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome1</td>\n",
       "      <td>1895035</td>\n",
       "      <td>1895057</td>\n",
       "      <td>22</td>\n",
       "      <td>nonref</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>d7669e06-6916-435d-9925-6c207b9258ee,fe42f608-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome1</td>\n",
       "      <td>18795334</td>\n",
       "      <td>18799060</td>\n",
       "      <td>3726</td>\n",
       "      <td>ref</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>5</td>\n",
       "      <td>c81615f3-be5b-4c2b-9b60-bad3cfe7feb4,a597c928-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome2</td>\n",
       "      <td>24991939</td>\n",
       "      <td>24991944</td>\n",
       "      <td>5</td>\n",
       "      <td>nonref</td>\n",
       "      <td>1</td>\n",
       "      <td>aa99dd0d-ba11-4f47-a79e-5dfedc9da5e2</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome2</td>\n",
       "      <td>30539627</td>\n",
       "      <td>30544358</td>\n",
       "      <td>4731</td>\n",
       "      <td>ref</td>\n",
       "      <td>1</td>\n",
       "      <td>b647706c-82df-4164-9168-4b200501ca6d</td>\n",
       "      <td>1</td>\n",
       "      <td>60b422a4-62dd-4740-bb99-d3de14200d9a</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome3</td>\n",
       "      <td>27718081</td>\n",
       "      <td>27722607</td>\n",
       "      <td>4526</td>\n",
       "      <td>ref</td>\n",
       "      <td>1</td>\n",
       "      <td>6c07ed7c-c720-49ac-9e4b-b83309306bae</td>\n",
       "      <td>1</td>\n",
       "      <td>f374ba15-f870-4ad2-9daf-ff983b2a8970</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome3</td>\n",
       "      <td>35158725</td>\n",
       "      <td>35158723</td>\n",
       "      <td>-2</td>\n",
       "      <td>nonref</td>\n",
       "      <td>1</td>\n",
       "      <td>cf60c0dc-a60a-4a67-acb7-28845f6eeb4c</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome3</td>\n",
       "      <td>41159456</td>\n",
       "      <td>41164183</td>\n",
       "      <td>4727</td>\n",
       "      <td>ref</td>\n",
       "      <td>3</td>\n",
       "      <td>c89a6c1c-495b-4c98-bcb1-376e6b1b3fb4,7e4c3981-...</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome4</td>\n",
       "      <td>4759534</td>\n",
       "      <td>4759532</td>\n",
       "      <td>-2</td>\n",
       "      <td>nonref</td>\n",
       "      <td>1</td>\n",
       "      <td>941656fa-26ed-4578-8aab-59824b70f133</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome4</td>\n",
       "      <td>35867416</td>\n",
       "      <td>35867451</td>\n",
       "      <td>35</td>\n",
       "      <td>nonref</td>\n",
       "      <td>1</td>\n",
       "      <td>a7a007a4-c486-4df5-854d-69f74d61d3d1</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2</td>\n",
       "      <td>chromosome4</td>\n",
       "      <td>26247770</td>\n",
       "      <td>26247794</td>\n",
       "      <td>24</td>\n",
       "      <td>nonref</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>b3e50c08-b937-486c-97ff-2ed57cab1f95,47eb892f-...</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome5</td>\n",
       "      <td>8967214</td>\n",
       "      <td>8971940</td>\n",
       "      <td>4726</td>\n",
       "      <td>ref</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1df3640c-8f38-4c47-a30c-b81b4c1c0a35</td>\n",
       "      <td>7</td>\n",
       "      <td>02a6bc39-ecc1-41f0-b5ca-c26c015ff4b1,84ab4f73-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome5</td>\n",
       "      <td>34862567</td>\n",
       "      <td>34862563</td>\n",
       "      <td>-4</td>\n",
       "      <td>nonref</td>\n",
       "      <td>1</td>\n",
       "      <td>dc8aa270-492a-4c08-933a-5c3b8eaa980e</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome5</td>\n",
       "      <td>41745603</td>\n",
       "      <td>41745600</td>\n",
       "      <td>-3</td>\n",
       "      <td>nonref</td>\n",
       "      <td>1</td>\n",
       "      <td>d69243ea-29eb-45f5-9a67-cb1adf455e12</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome6</td>\n",
       "      <td>6237082</td>\n",
       "      <td>6237101</td>\n",
       "      <td>19</td>\n",
       "      <td>nonref</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>a4c02692-8649-450c-87ec-9d8d532f47cf</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome7</td>\n",
       "      <td>12013941</td>\n",
       "      <td>12013940</td>\n",
       "      <td>-1</td>\n",
       "      <td>nonref</td>\n",
       "      <td>1</td>\n",
       "      <td>92e0de8b-6882-4283-87a2-491864754ad3</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome7</td>\n",
       "      <td>13174817</td>\n",
       "      <td>13174813</td>\n",
       "      <td>-4</td>\n",
       "      <td>nonref</td>\n",
       "      <td>1</td>\n",
       "      <td>29af0fb0-6c1a-44d8-8425-57c5f585641d</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome8</td>\n",
       "      <td>11878585</td>\n",
       "      <td>11878581</td>\n",
       "      <td>-4</td>\n",
       "      <td>nonref</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>114221bc-ac48-4940-8416-0ba3ab107414,dc28601b-...</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome8</td>\n",
       "      <td>22629055</td>\n",
       "      <td>22629051</td>\n",
       "      <td>-4</td>\n",
       "      <td>nonref</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>75a9a3f3-13e0-4403-a2bb-a1e6763fdadf</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome8</td>\n",
       "      <td>31372224</td>\n",
       "      <td>31372250</td>\n",
       "      <td>26</td>\n",
       "      <td>nonref</td>\n",
       "      <td>1</td>\n",
       "      <td>ac77f568-5294-480f-8ed6-410cdc82fdde</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome9</td>\n",
       "      <td>9220466</td>\n",
       "      <td>9226437</td>\n",
       "      <td>5971</td>\n",
       "      <td>ref</td>\n",
       "      <td>2</td>\n",
       "      <td>070100af-93c2-4c1c-b426-0e6852ff5740,ba6c24b6-...</td>\n",
       "      <td>1</td>\n",
       "      <td>f1775b2d-b6a5-490c-9f57-745df380c634</td>\n",
       "      <td>8</td>\n",
       "      <td>4fcac5a2-484b-484f-b76c-5702d173611d,dd74e399-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome9</td>\n",
       "      <td>27459308</td>\n",
       "      <td>27463311</td>\n",
       "      <td>4003</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>5ff9d997-242d-4ea2-9b0d-d0211cf20736</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome9</td>\n",
       "      <td>28529396</td>\n",
       "      <td>28534569</td>\n",
       "      <td>5173</td>\n",
       "      <td>ref</td>\n",
       "      <td>4</td>\n",
       "      <td>c20f1362-fbcb-4230-8a32-73659bc06318,10e616cd-...</td>\n",
       "      <td>1</td>\n",
       "      <td>fe5cf5ad-f0e9-4948-8fd2-796e936c15c3</td>\n",
       "      <td>4</td>\n",
       "      <td>d3f24ffe-7781-41a2-ad5c-2105e3cca0c8,068da520-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1</td>\n",
       "      <td>chromosome9</td>\n",
       "      <td>31434581</td>\n",
       "      <td>31434577</td>\n",
       "      <td>-4</td>\n",
       "      <td>nonref</td>\n",
       "      <td>1</td>\n",
       "      <td>66e7bd3c-b6e7-476e-98f7-79d006377107</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    pkt   chromosome     start      stop  length   is_ref  left_K10B_num  \\\n",
       "0     1  chromosome1   1895035   1895057      22   nonref              0   \n",
       "1     1  chromosome1  18795334  18799060    3726      ref              0   \n",
       "2     1  chromosome2  24991939  24991944       5   nonref              1   \n",
       "3     1  chromosome2  30539627  30544358    4731      ref              1   \n",
       "4     1  chromosome3  27718081  27722607    4526      ref              1   \n",
       "5     1  chromosome3  35158725  35158723      -2   nonref              1   \n",
       "6     1  chromosome3  41159456  41164183    4727      ref              3   \n",
       "7     1  chromosome4   4759534   4759532      -2   nonref              1   \n",
       "8     1  chromosome4  35867416  35867451      35   nonref              1   \n",
       "9     2  chromosome4  26247770  26247794      24   nonref              0   \n",
       "10    1  chromosome5   8967214   8971940    4726      ref              0   \n",
       "11    1  chromosome5  34862567  34862563      -4   nonref              1   \n",
       "12    1  chromosome5  41745603  41745600      -3   nonref              1   \n",
       "13    1  chromosome6   6237082   6237101      19   nonref              0   \n",
       "14    1  chromosome7  12013941  12013940      -1   nonref              1   \n",
       "15    1  chromosome7  13174817  13174813      -4   nonref              1   \n",
       "16    1  chromosome8  11878585  11878581      -4   nonref              0   \n",
       "17    1  chromosome8  22629055  22629051      -4   nonref              0   \n",
       "18    1  chromosome8  31372224  31372250      26   nonref              1   \n",
       "19    1  chromosome9   9220466   9226437    5971      ref              2   \n",
       "20    1  chromosome9  27459308  27463311    4003  unknown              1   \n",
       "21    1  chromosome9  28529396  28534569    5173      ref              4   \n",
       "22    1  chromosome9  31434581  31434577      -4   nonref              1   \n",
       "\n",
       "                                        left_K10B_rec  left_K10f_num  \\\n",
       "0                                                                  0   \n",
       "1                                                                  0   \n",
       "2                aa99dd0d-ba11-4f47-a79e-5dfedc9da5e2              0   \n",
       "3                b647706c-82df-4164-9168-4b200501ca6d              1   \n",
       "4                6c07ed7c-c720-49ac-9e4b-b83309306bae              1   \n",
       "5                cf60c0dc-a60a-4a67-acb7-28845f6eeb4c              0   \n",
       "6   c89a6c1c-495b-4c98-bcb1-376e6b1b3fb4,7e4c3981-...              0   \n",
       "7                941656fa-26ed-4578-8aab-59824b70f133              0   \n",
       "8                a7a007a4-c486-4df5-854d-69f74d61d3d1              0   \n",
       "9                                                                  2   \n",
       "10                                                                 1   \n",
       "11               dc8aa270-492a-4c08-933a-5c3b8eaa980e              0   \n",
       "12               d69243ea-29eb-45f5-9a67-cb1adf455e12              0   \n",
       "13                                                                 1   \n",
       "14               92e0de8b-6882-4283-87a2-491864754ad3              0   \n",
       "15               29af0fb0-6c1a-44d8-8425-57c5f585641d              0   \n",
       "16                                                                 2   \n",
       "17                                                                 1   \n",
       "18               ac77f568-5294-480f-8ed6-410cdc82fdde              0   \n",
       "19  070100af-93c2-4c1c-b426-0e6852ff5740,ba6c24b6-...              1   \n",
       "20               5ff9d997-242d-4ea2-9b0d-d0211cf20736              0   \n",
       "21  c20f1362-fbcb-4230-8a32-73659bc06318,10e616cd-...              1   \n",
       "22               66e7bd3c-b6e7-476e-98f7-79d006377107              0   \n",
       "\n",
       "                                        left_K10f_rec  left_LK5_1_3_num  \\\n",
       "0                                                                     3   \n",
       "1                                                                     5   \n",
       "2                                                                     0   \n",
       "3                60b422a4-62dd-4740-bb99-d3de14200d9a                 0   \n",
       "4                f374ba15-f870-4ad2-9daf-ff983b2a8970                 0   \n",
       "5                                                                     0   \n",
       "6                                                                     0   \n",
       "7                                                                     0   \n",
       "8                                                                     0   \n",
       "9   b3e50c08-b937-486c-97ff-2ed57cab1f95,47eb892f-...                 0   \n",
       "10               1df3640c-8f38-4c47-a30c-b81b4c1c0a35                 7   \n",
       "11                                                                    0   \n",
       "12                                                                    0   \n",
       "13               a4c02692-8649-450c-87ec-9d8d532f47cf                 0   \n",
       "14                                                                    0   \n",
       "15                                                                    0   \n",
       "16  114221bc-ac48-4940-8416-0ba3ab107414,dc28601b-...                 0   \n",
       "17               75a9a3f3-13e0-4403-a2bb-a1e6763fdadf                 0   \n",
       "18                                                                    0   \n",
       "19               f1775b2d-b6a5-490c-9f57-745df380c634                 8   \n",
       "20                                                                    0   \n",
       "21               fe5cf5ad-f0e9-4948-8fd2-796e936c15c3                 4   \n",
       "22                                                                    0   \n",
       "\n",
       "                                     left_LK5_1_3_rec  \n",
       "0   d7669e06-6916-435d-9925-6c207b9258ee,fe42f608-...  \n",
       "1   c81615f3-be5b-4c2b-9b60-bad3cfe7feb4,a597c928-...  \n",
       "2                                                      \n",
       "3                                                      \n",
       "4                                                      \n",
       "5                                                      \n",
       "6                                                      \n",
       "7                                                      \n",
       "8                                                      \n",
       "9                                                      \n",
       "10  02a6bc39-ecc1-41f0-b5ca-c26c015ff4b1,84ab4f73-...  \n",
       "11                                                     \n",
       "12                                                     \n",
       "13                                                     \n",
       "14                                                     \n",
       "15                                                     \n",
       "16                                                     \n",
       "17                                                     \n",
       "18                                                     \n",
       "19  4fcac5a2-484b-484f-b76c-5702d173611d,dd74e399-...  \n",
       "20                                                     \n",
       "21  d3f24ffe-7781-41a2-ad5c-2105e3cca0c8,068da520-...  \n",
       "22                                                     "
      ]
     },
     "execution_count": 409,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xxx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "7231be0e-aeae-478f-b465-ec8f9d4a4843",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 from 0.0 to 9.92\n",
      "2 from 10.23 to 15.19\n",
      "3 from 15.5 to 22.63\n",
      "4 from 22.94 to 27.59\n",
      "5 from 27.900000000000002 to 30.07\n",
      "6 from 30.38 to 31.0\n"
     ]
    }
   ],
   "source": [
    "x = 31\n",
    "print(1, f\"from {0/100*x} to {32/100*x}\")\n",
    "print(2, f\"from {33/100*x} to {49/100*x}\")\n",
    "print(3, f\"from {50/100*x} to {73/100*x}\")\n",
    "print(4, f\"from {74/100*x} to {89/100*x}\")\n",
    "print(5, f\"from {90/100*x} to {97/100*x}\")\n",
    "print(6, f\"from {98/100*x} to {100/100*x}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2387a5eb-dd07-4b9f-a745-f474a8213055",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
