{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8aa228f6-1f9e-446b-bd3a-944ac913bc3e",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b56d31e2-34ca-480e-a32b-d70d6e6a3031",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7fdf9c57-bfb4-44c7-ac99-7a581de474fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from bioinfokit.analys import Fasta\n",
    "import os\n",
    "# helpers function\n",
    "\n",
    "def get_fasta(filename):\n",
    "    n = 0\n",
    "    sequence = \"\"\n",
    "    te_name = \"\"\n",
    "    with open(filename) as f:\n",
    "        for line in f.readlines():\n",
    "            if line.startswith('>'):\n",
    "                n = n + 1\n",
    "                if not n == 1:\n",
    "                    first_n = False\n",
    "                    yield te_name, sequence\n",
    "                te_name = line[1:].strip()\n",
    "                sequence = \"\"\n",
    "            else:\n",
    "                if te_name:\n",
    "                    sequence = sequence + line.strip()\n",
    "    yield te_name, sequence\n",
    "\n",
    "def get_len_fasta(filename):\n",
    "    counter = {}\n",
    "    with open(filename) as f:\n",
    "        te_name = \"\"\n",
    "        for line in f.readlines():\n",
    "            if line.startswith('>'):\n",
    "                te_name = line[1:].strip()\n",
    "                counter[te_name] = 0\n",
    "            else:\n",
    "                if te_name:\n",
    "                    counter[te_name] = counter[te_name] + len(line.strip()) - line.count(\"~\")\n",
    "    return counter\n",
    "\n",
    "def get_from_te(fasta, te_name):\n",
    "    te_len = get_len_fasta(fasta)[te_name]\n",
    "    return int(te_len - te_len * dif_percent / 100)\n",
    "\n",
    "def get_to_te(fasta, te_name):\n",
    "    te_len = get_len_fasta(fasta)[te_name]\n",
    "    return int(te_len + te_len * dif_percent / 100)\n",
    "\n",
    "def read_fasta(filename):\n",
    "    return pd.read_csv(filename, sep=\"\\t\", header=None).rename(headers_dict, axis=1)\n",
    "\n",
    "def save_fasta(df, filename):\n",
    "    df.to_csv(filename, sep='\\t', index=False, encoding='utf-8', header=False)\n",
    "\n",
    "def _create_part_raport(df):\n",
    "    if len(df) == 0:\n",
    "        return pd.DataFrame(\n",
    "            {\n",
    "                'qseqid': [],\n",
    "                'sseqid': [],\n",
    "                'start_te': [],\n",
    "                'end_te': [],\n",
    "                'start_genome_te': [],\n",
    "                'end_genome_te': [],\n",
    "    \n",
    "                \n",
    "                \n",
    "                'qstart1': [],\n",
    "                'qend1': [],\n",
    "                'sstart1': [],\n",
    "                'send1': [],\n",
    "                \n",
    "                'qstart2': [],\n",
    "                'qend2': [],\n",
    "                'sstart2': [],\n",
    "                'send2': [],\n",
    "    \n",
    "                'length': [],\n",
    "                'length_genome': [],\n",
    "            }\n",
    "        )\n",
    "    # there is one qseqid\n",
    "    # it should be sortet to q_start\n",
    "    df = df.sort_values(by=['sseqid', 'qstart', 'qend', 'sstart', 'send'])\n",
    "    start = df[:-1].reset_index(drop=True)\n",
    "    end = df[1:].reset_index(drop=True)\n",
    "    return pd.DataFrame(\n",
    "        {\n",
    "            'qseqid':df.qseqid.values[0],\n",
    "            'sseqid':df.sseqid.values[0],\n",
    "            'start_te': start.qend,\n",
    "            'end_te': end.qstart,\n",
    "            'start_genome_te': start.send,\n",
    "            'end_genome_te': end.sstart,\n",
    "\n",
    "            \n",
    "            \n",
    "            'qstart1': end.qstart,\n",
    "            'qend1': end.qend,\n",
    "            'sstart1': end.sstart,\n",
    "            'send1': end.send,\n",
    "            \n",
    "            'qstart2': start.qstart,\n",
    "            'qend2': start.qend,\n",
    "            'sstart2': start.sstart,\n",
    "            'send2': start.send,\n",
    "\n",
    "            'length': end.qstart - start.qend,\n",
    "            'length_genome': end.sstart - start.send,\n",
    "        }\n",
    "    )\n",
    "\n",
    "\n",
    "def create_report(df):\n",
    "    series_list = []\n",
    "    for qseqid in df.qseqid.unique():\n",
    "        okno = df[df.qseqid == qseqid]\n",
    "        for sseqid in okno.sseqid.unique():\n",
    "            okno2 = okno[(okno.sseqid == sseqid)]\n",
    "            series_list.append(_create_part_raport(okno2))\n",
    "    return pd.concat(series_list) if series_list else _create_part_raport(df)\n",
    "\n",
    "def filter_finish(df):\n",
    "    return df[\n",
    "        (df.length < to_te)\n",
    "        & (df.length > from_te)\n",
    "        & (df.length_genome > -100)\n",
    "        & (df.length_genome < 12000)\n",
    "    ]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f5e52fe-0279-45b3-9ac0-fbc3729aa5ac",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fe2261af-9c56-4400-aeab-d929eb4d76c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ont = [\"K10B\", \"K10f\", \"LK5_1_3\"]\n",
    "ont = [\"K10B\", \"K10f\"]\n",
    "ont = [\"konik4\"]\n",
    "genome_path = \"genome/genomic2.fna\"\n",
    "te_path = \"te/Alex1_3.fas\"\n",
    "dif_percent = 30\n",
    "\n",
    "te_name = 'Alex1_chr1:18795335-18799058'\n",
    "te_name = 'Alex1'\n",
    "te_len = get_len_fasta(te_path)[te_name]\n",
    "from_te = get_from_te(te_path, te_name)\n",
    "to_te = get_to_te(te_path, te_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3b790fca-e1b9-4207-bcb9-e4ace324321d",
   "metadata": {},
   "outputs": [],
   "source": [
    "headers = ['qseqid', 'sseqid',  'pident', 'length', 'mismatch', 'gapopen', 'qstart', 'qend', 'sstart', 'send', 'expect value', 'bitscore']\n",
    "headers_dict = {i[0]:i[1] for i in enumerate(headers)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecd1cc04-aba2-42fd-b345-a1cecd4f6aa6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "1\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "2\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "3\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "4\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "5\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "6\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "7\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "8\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "9\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "10\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "11\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n",
      "12\n",
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "def rand_split(sequence, splits):\n",
    "    for splitLen in splits:\n",
    "        if splitLen > len(sequence):\n",
    "            break\n",
    "        yield sequence[:splitLen]\n",
    "        sequence = sequence[splitLen::]\n",
    "\n",
    "def rand_gen(int_list):\n",
    "    while True:\n",
    "        yield int(random.choice(int_list))\n",
    "\n",
    "def cut_genome(genome_path, ont_path, coverage, filename):\n",
    "    possibility_len = list(get_len_fasta(ont_path).values())\n",
    "    n = 0\n",
    "    for i in range(coverage):\n",
    "        print(i)\n",
    "        for name, genome_sequence in get_fasta(genome_path):\n",
    "            print(name)\n",
    "            with open(filename, 'a') as file:\n",
    "                for seq in rand_split(genome_sequence, rand_gen(possibility_len)):\n",
    "                    n = n + 1\n",
    "                    file.write(f\">{name}_{i}_{n}\\n{seq}\\n\")\n",
    "cut_genome(genome_path, \"ont/K10f.fasta\", 20, 'cutted_genome.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "3fa43c27-6370-4465-8e1a-4ad578831a87",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosome1\n",
      "chromosome2\n",
      "chromosome3\n",
      "chromosome4\n",
      "chromosome5\n",
      "chromosome6\n",
      "chromosome7\n",
      "chromosome8\n",
      "chromosome9\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "w = 2\n",
    "okno = list(okno)\n",
    "n = 0\n",
    "coverage = 1\n",
    "for i in range(coverage):\n",
    "    for name, genome_sequence in get_fasta(genome_path):\n",
    "        print(name)\n",
    "        with open(f'migi{w}.mi', 'a') as file:\n",
    "            for seq in rand_split(genome_sequence, rand_gen([30000])):\n",
    "                n = n + 1\n",
    "                file.write(f\">{name}_{w}_{n}\\n\")\n",
    "                file.write(seq)\n",
    "                file.write(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a8be1029-7aca-4774-a0fd-1601ab075298",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['123', '456', '789', '0']\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "674381f7-965c-4216-a694-5d30dda82d1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 04/17/2024 05:14:20\n",
      "New DB name:   /home/kiki/data2/genome/genome\n",
      "New DB title:  genome/genomic2.fna\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /home/kiki/data2/genome/genome\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 9 sequences in 5.50287 seconds.\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n"
     ]
    }
   ],
   "source": [
    "def mask_genome(te_name, te_path, genome_path, make_db=False):\n",
    "    os.system(f\"mkdir -p masked_genome\")\n",
    "    if make_db:\n",
    "        os.system(f\"makeblastdb -in {genome_path} -dbtype nucl -out genome/genome\")\n",
    "    os.system(f\"blastn -num_threads 20 -outfmt 6 -query {te_path} -db genome/genome -out masked_genome/{te_name}_genome.bl -perc_identity 0.9\")\n",
    "    tex = read_fasta(f\"masked_genome/{te_name}_genome.bl\")\n",
    "    tex = tex[tex.qseqid == te_name]\n",
    "    tex[\"start\"] = tex['sstart'].where(tex['sstart'] <= tex[\"send\"], other=tex['send'])\n",
    "    tex[\"stop\"] = tex['sstart'].where(tex['sstart'] > tex[\"send\"], other=tex['send'])\n",
    "    # todo\n",
    "    save_fasta(tex[[\"sseqid\", \"start\", \"stop\"]], f\"masked_genome/to_mask_genome_{te_name}.txt\")\n",
    "    plus_tex = tex[tex.sstart < tex.send]\n",
    "    minus_tex = tex[tex.sstart > tex.send]\n",
    "\n",
    "    merge_headers = [\"sseqid\", \"sstart\", \"send\"]\n",
    "    save_fasta(plus_tex[merge_headers].sort_values(by=merge_headers), f\"masked_genome/plus_to_mask_genome_{te_name}.bed\")\n",
    "    merge_headers = [\"sseqid\", \"send\", \"sstart\"]\n",
    "    save_fasta(minus_tex[merge_headers].sort_values(by=merge_headers), f\"masked_genome/minus_to_mask_genome_{te_name}.bed\")\n",
    "\n",
    "    os.system(f\"bedtools merge -i masked_genome/plus_to_mask_genome_{te_name}.bed -d 200 > masked_genome/plus_to_mask_genome_{te_name}.bed.merged\")\n",
    "    os.system(f\"bedtools merge -i masked_genome/minus_to_mask_genome_{te_name}.bed -d 200 > masked_genome/minus_to_mask_genome_{te_name}.bed.merged\")\n",
    "    minus_merged =  pd.read_csv(f\"masked_genome/minus_to_mask_genome_{te_name}.bed.merged\", sep=\"\\t\", header=None)\n",
    "    plus_merged =  pd.read_csv(f\"masked_genome/plus_to_mask_genome_{te_name}.bed.merged\", sep=\"\\t\", header=None)\n",
    "    merged_genome = pd.concat(\n",
    "        [\n",
    "            minus_merged[\n",
    "                (minus_merged[2] - minus_merged[1] > from_te)\n",
    "                & (minus_merged[2] - minus_merged[1] < to_te)\n",
    "            ],\n",
    "            plus_merged[\n",
    "                (plus_merged[2] - plus_merged[1] > from_te)\n",
    "                & (plus_merged[2] - plus_merged[1] < to_te)\n",
    "            ],\n",
    "        ]\n",
    "    )\n",
    "    merged_genome[\"length\"] = merged_genome[2] - merged_genome[1]\n",
    "    save_fasta(merged_genome, f\"masked_genome/to_mask_genome_{te_name}.bed.merged\")\n",
    "    os.system(f\"bedtools maskfasta -fi {genome_path} -bed masked_genome/to_mask_genome_{te_name}.bed.merged -fo genome/{te_name}_masked_genome\")\n",
    "    return merged_genome\n",
    "doxi = mask_genome(te_name, te_path, genome_path, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "0c11fa5f-93b2-470c-b2a0-da917515cb69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chromosome2</td>\n",
       "      <td>30539632</td>\n",
       "      <td>30544357</td>\n",
       "      <td>4725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chromosome3</td>\n",
       "      <td>27717880</td>\n",
       "      <td>27722606</td>\n",
       "      <td>4726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chromosome5</td>\n",
       "      <td>8967214</td>\n",
       "      <td>8971939</td>\n",
       "      <td>4725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chromosome9</td>\n",
       "      <td>9220466</td>\n",
       "      <td>9226148</td>\n",
       "      <td>5682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chromosome9</td>\n",
       "      <td>28529339</td>\n",
       "      <td>28534568</td>\n",
       "      <td>5229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome1</td>\n",
       "      <td>18795334</td>\n",
       "      <td>18799058</td>\n",
       "      <td>3724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chromosome3</td>\n",
       "      <td>41159456</td>\n",
       "      <td>41164182</td>\n",
       "      <td>4726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chromosome4</td>\n",
       "      <td>35827649</td>\n",
       "      <td>35832374</td>\n",
       "      <td>4725</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0         1         2  length\n",
       "1   chromosome2  30539632  30544357    4725\n",
       "4   chromosome3  27717880  27722606    4726\n",
       "6   chromosome5   8967214   8971939    4725\n",
       "9   chromosome9   9220466   9226148    5682\n",
       "10  chromosome9  28529339  28534568    5229\n",
       "0   chromosome1  18795334  18799058    3724\n",
       "2   chromosome3  41159456  41164182    4726\n",
       "5   chromosome4  35827649  35832374    4725"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "doxi.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "5a8dac83-918e-437a-b1d5-53817d3cea84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>start_te</th>\n",
       "      <th>end_te</th>\n",
       "      <th>start_genome_te</th>\n",
       "      <th>end_genome_te</th>\n",
       "      <th>qstart1</th>\n",
       "      <th>qend1</th>\n",
       "      <th>sstart1</th>\n",
       "      <th>send1</th>\n",
       "      <th>qstart2</th>\n",
       "      <th>qend2</th>\n",
       "      <th>sstart2</th>\n",
       "      <th>send2</th>\n",
       "      <th>length</th>\n",
       "      <th>length_genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome1_2_627:11335-23058</td>\n",
       "      <td>chromosome1</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>7725.0</td>\n",
       "      <td>18795334.0</td>\n",
       "      <td>18799059.0</td>\n",
       "      <td>7725.0</td>\n",
       "      <td>11724.0</td>\n",
       "      <td>18799059.0</td>\n",
       "      <td>18803058.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>18791335.0</td>\n",
       "      <td>18795334.0</td>\n",
       "      <td>3725.0</td>\n",
       "      <td>3725.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome2_1_3280:10633-23357</td>\n",
       "      <td>chromosome2</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>8726.0</td>\n",
       "      <td>30539632.0</td>\n",
       "      <td>30544358.0</td>\n",
       "      <td>8726.0</td>\n",
       "      <td>12725.0</td>\n",
       "      <td>30544358.0</td>\n",
       "      <td>30548357.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>30535633.0</td>\n",
       "      <td>30539632.0</td>\n",
       "      <td>4726.0</td>\n",
       "      <td>4726.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome3_1_4923:13881-25000</td>\n",
       "      <td>chromosome3</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>8727.0</td>\n",
       "      <td>27717880.0</td>\n",
       "      <td>27722607.0</td>\n",
       "      <td>8727.0</td>\n",
       "      <td>11120.0</td>\n",
       "      <td>27722607.0</td>\n",
       "      <td>27725000.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>27713881.0</td>\n",
       "      <td>27717880.0</td>\n",
       "      <td>4727.0</td>\n",
       "      <td>4727.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome3_1_5461:5457-18182</td>\n",
       "      <td>chromosome3</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>8727.0</td>\n",
       "      <td>41159456.0</td>\n",
       "      <td>41164183.0</td>\n",
       "      <td>8727.0</td>\n",
       "      <td>12726.0</td>\n",
       "      <td>41164183.0</td>\n",
       "      <td>41168182.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>41155457.0</td>\n",
       "      <td>41159456.0</td>\n",
       "      <td>4727.0</td>\n",
       "      <td>4727.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome4_0_9075:3650-16374</td>\n",
       "      <td>chromosome4</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>8726.0</td>\n",
       "      <td>35827649.0</td>\n",
       "      <td>35832375.0</td>\n",
       "      <td>8726.0</td>\n",
       "      <td>12725.0</td>\n",
       "      <td>35832375.0</td>\n",
       "      <td>35836374.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>35823650.0</td>\n",
       "      <td>35827649.0</td>\n",
       "      <td>4726.0</td>\n",
       "      <td>4726.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome4_1_7260:1-11374</td>\n",
       "      <td>chromosome4</td>\n",
       "      <td>2649.0</td>\n",
       "      <td>7375.0</td>\n",
       "      <td>35827649.0</td>\n",
       "      <td>35832375.0</td>\n",
       "      <td>7375.0</td>\n",
       "      <td>11374.0</td>\n",
       "      <td>35832375.0</td>\n",
       "      <td>35836374.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2649.0</td>\n",
       "      <td>35825001.0</td>\n",
       "      <td>35827649.0</td>\n",
       "      <td>4726.0</td>\n",
       "      <td>4726.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome4_2_6050:3650-16374</td>\n",
       "      <td>chromosome4</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>8726.0</td>\n",
       "      <td>35827649.0</td>\n",
       "      <td>35832375.0</td>\n",
       "      <td>8726.0</td>\n",
       "      <td>12725.0</td>\n",
       "      <td>35832375.0</td>\n",
       "      <td>35836374.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>35823650.0</td>\n",
       "      <td>35827649.0</td>\n",
       "      <td>4726.0</td>\n",
       "      <td>4726.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chromosome5_0_9528:3215-15939</td>\n",
       "      <td>chromosome5</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>8726.0</td>\n",
       "      <td>8967214.0</td>\n",
       "      <td>8971940.0</td>\n",
       "      <td>8726.0</td>\n",
       "      <td>12725.0</td>\n",
       "      <td>8971940.0</td>\n",
       "      <td>8975939.0</td>\n",
       "      <td>2094.0</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>8965308.0</td>\n",
       "      <td>8967214.0</td>\n",
       "      <td>4726.0</td>\n",
       "      <td>4726.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chromosome5_1_7621:13215-25000</td>\n",
       "      <td>chromosome5</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>8726.0</td>\n",
       "      <td>8967214.0</td>\n",
       "      <td>8971940.0</td>\n",
       "      <td>8726.0</td>\n",
       "      <td>11786.0</td>\n",
       "      <td>8971940.0</td>\n",
       "      <td>8975000.0</td>\n",
       "      <td>2094.0</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>8965308.0</td>\n",
       "      <td>8967214.0</td>\n",
       "      <td>4726.0</td>\n",
       "      <td>4726.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome9_2_11247:6467-20148</td>\n",
       "      <td>chromosome9</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>9683.0</td>\n",
       "      <td>9220466.0</td>\n",
       "      <td>9226149.0</td>\n",
       "      <td>9683.0</td>\n",
       "      <td>13682.0</td>\n",
       "      <td>9226149.0</td>\n",
       "      <td>9230148.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>9216467.0</td>\n",
       "      <td>9220466.0</td>\n",
       "      <td>5683.0</td>\n",
       "      <td>5683.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome9_0_17837:5340-18568</td>\n",
       "      <td>chromosome9</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>9230.0</td>\n",
       "      <td>28529339.0</td>\n",
       "      <td>28534569.0</td>\n",
       "      <td>9230.0</td>\n",
       "      <td>13229.0</td>\n",
       "      <td>28534569.0</td>\n",
       "      <td>28538568.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>28525340.0</td>\n",
       "      <td>28529339.0</td>\n",
       "      <td>5230.0</td>\n",
       "      <td>5230.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome9_1_14269:340-13568</td>\n",
       "      <td>chromosome9</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>9230.0</td>\n",
       "      <td>28529339.0</td>\n",
       "      <td>28534569.0</td>\n",
       "      <td>9230.0</td>\n",
       "      <td>13229.0</td>\n",
       "      <td>28534569.0</td>\n",
       "      <td>28538568.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>28525340.0</td>\n",
       "      <td>28529339.0</td>\n",
       "      <td>5230.0</td>\n",
       "      <td>5230.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           qseqid       sseqid  start_te  end_te  \\\n",
       "0   chromosome1_2_627:11335-23058  chromosome1    4000.0  7725.0   \n",
       "0  chromosome2_1_3280:10633-23357  chromosome2    4000.0  8726.0   \n",
       "0  chromosome3_1_4923:13881-25000  chromosome3    4000.0  8727.0   \n",
       "0   chromosome3_1_5461:5457-18182  chromosome3    4000.0  8727.0   \n",
       "0   chromosome4_0_9075:3650-16374  chromosome4    4000.0  8726.0   \n",
       "0      chromosome4_1_7260:1-11374  chromosome4    2649.0  7375.0   \n",
       "0   chromosome4_2_6050:3650-16374  chromosome4    4000.0  8726.0   \n",
       "1   chromosome5_0_9528:3215-15939  chromosome5    4000.0  8726.0   \n",
       "1  chromosome5_1_7621:13215-25000  chromosome5    4000.0  8726.0   \n",
       "0  chromosome9_2_11247:6467-20148  chromosome9    4000.0  9683.0   \n",
       "0  chromosome9_0_17837:5340-18568  chromosome9    4000.0  9230.0   \n",
       "0   chromosome9_1_14269:340-13568  chromosome9    4000.0  9230.0   \n",
       "\n",
       "   start_genome_te  end_genome_te  qstart1    qend1     sstart1       send1  \\\n",
       "0       18795334.0     18799059.0   7725.0  11724.0  18799059.0  18803058.0   \n",
       "0       30539632.0     30544358.0   8726.0  12725.0  30544358.0  30548357.0   \n",
       "0       27717880.0     27722607.0   8727.0  11120.0  27722607.0  27725000.0   \n",
       "0       41159456.0     41164183.0   8727.0  12726.0  41164183.0  41168182.0   \n",
       "0       35827649.0     35832375.0   8726.0  12725.0  35832375.0  35836374.0   \n",
       "0       35827649.0     35832375.0   7375.0  11374.0  35832375.0  35836374.0   \n",
       "0       35827649.0     35832375.0   8726.0  12725.0  35832375.0  35836374.0   \n",
       "1        8967214.0      8971940.0   8726.0  12725.0   8971940.0   8975939.0   \n",
       "1        8967214.0      8971940.0   8726.0  11786.0   8971940.0   8975000.0   \n",
       "0        9220466.0      9226149.0   9683.0  13682.0   9226149.0   9230148.0   \n",
       "0       28529339.0     28534569.0   9230.0  13229.0  28534569.0  28538568.0   \n",
       "0       28529339.0     28534569.0   9230.0  13229.0  28534569.0  28538568.0   \n",
       "\n",
       "   qstart2   qend2     sstart2       send2  length  length_genome  \n",
       "0      1.0  4000.0  18791335.0  18795334.0  3725.0         3725.0  \n",
       "0      1.0  4000.0  30535633.0  30539632.0  4726.0         4726.0  \n",
       "0      1.0  4000.0  27713881.0  27717880.0  4727.0         4727.0  \n",
       "0      1.0  4000.0  41155457.0  41159456.0  4727.0         4727.0  \n",
       "0      1.0  4000.0  35823650.0  35827649.0  4726.0         4726.0  \n",
       "0      1.0  2649.0  35825001.0  35827649.0  4726.0         4726.0  \n",
       "0      1.0  4000.0  35823650.0  35827649.0  4726.0         4726.0  \n",
       "1   2094.0  4000.0   8965308.0   8967214.0  4726.0         4726.0  \n",
       "1   2094.0  4000.0   8965308.0   8967214.0  4726.0         4726.0  \n",
       "0      1.0  4000.0   9216467.0   9220466.0  5683.0         5683.0  \n",
       "0      1.0  4000.0  28525340.0  28529339.0  5230.0         5230.0  \n",
       "0      1.0  4000.0  28525340.0  28529339.0  5230.0         5230.0  "
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "oxi.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6b7182f8-9ecc-43d9-ba3e-b7e7949eccfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "def first_part():\n",
    "    headers = ['qseqid', 'sseqid',  'pident', 'length', 'mismatch', 'gapopen', 'qstart', 'qend', 'sstart', 'send', 'expect value', 'bitscore']\n",
    "    headers_dict = {i[0]:i[1] for i in enumerate(headers)}\n",
    "    os.system(\"mkdir -p blast\")\n",
    "    os.system(\"mkdir -p first\")\n",
    "    if True:\n",
    "        #for file in ont:\n",
    "        #    os.system(f\"seqtk seq -a ont/{file}.fastq.gz > ont/{file}.fasta\")\n",
    "        #!seqtk seq -a ont/K10B.fastq.gz > ont/K10B.fasta\n",
    "        for file in ont:\n",
    "            os.system(f\"makeblastdb -in ont/{file}.fasta -dbtype nucl -out ont/{file}\")\n",
    "        #!makeblastdb -in ont/K10B.fasta -dbtype nucl -out ont/K10B\n",
    "        for file in ont:\n",
    "            os.system(f\"blastn -num_threads 20 -outfmt 6 -query {te_path} -db ont/{file} -out blast/TE_{file}.bl -dust no -perc_identity 0.9\")\n",
    "        #!blastn -num_threads 20 -outfmt 6 -query te/Alex1_Alex3.fasta -db ont/K10B -out blast/Alex_K10B.bl -dust no -perc_identity 0.9\n",
    "\n",
    "    for o in ont:\n",
    "        te_ont_blast = read_fasta(f\"blast/TE_{o}.bl\")\n",
    "    \n",
    "        alex1 = te_ont_blast[te_ont_blast.qseqid.str.contains('Alex1')]\n",
    "        plus_alex1 = alex1[alex1.sstart < alex1.send]\n",
    "        minus_alex1 = alex1[alex1.sstart > alex1.send]\n",
    "    \n",
    "    \n",
    "        p_bed_df = plus_alex1[['sseqid', 'sstart', 'send', 'qseqid']].sort_values(by=['sseqid', 'sstart', 'send'])\n",
    "        m_bed_df = minus_alex1[['sseqid', 'send', 'sstart', 'qseqid']].sort_values(by=['sseqid', 'send', 'sstart'])\n",
    "    \n",
    "        save_fasta(m_bed_df, f'first/m_raw_first_blast_{o}.bed')\n",
    "        save_fasta(p_bed_df, f'first/p_raw_first_blast_{o}.bed')\n",
    "\n",
    "        os.system(f\"bedtools merge -i first/p_raw_first_blast_{o}.bed -d 1500 -c 4 -o collapse  > first/p_first_blast_{o}.bed.merged\")\n",
    "        os.system(f\"bedtools merge -i first/m_raw_first_blast_{o}.bed -d 1500 -c 4 -o collapse  > first/m_first_blast_{o}.bed.merged\")\n",
    "\n",
    "        print(\"first part\")\n",
    "    \n",
    "        p_merged = pd.read_csv(\n",
    "            f\"first/p_first_blast_{o}.bed.merged\", sep=\"\\t\", header=None).rename(\n",
    "            {0:'sseqid', 1:'sstart', 2:'send', 3:'qseqid'}, axis=1\n",
    "        )\n",
    "        m_merged = pd.read_csv(\n",
    "            f\"first/m_first_blast_{o}.bed.merged\", sep=\"\\t\", header=None).rename(\n",
    "            {0:'sseqid', 1:'sstart', 2:'send', 3:'qseqid'}, axis=1\n",
    "        )\n",
    "    \n",
    "        p_filtered_first_bed = p_merged[(p_merged.send - p_merged.sstart) > from_te]\n",
    "        m_filtered_first_bed = m_merged[(m_merged.send - m_merged.sstart) > from_te]\n",
    "    \n",
    "        pd.options.mode.copy_on_write = True\n",
    "        p_filtered_first_bed['end'] = p_filtered_first_bed['send'] + 4000\n",
    "        p_filtered_first_bed['start'] = p_filtered_first_bed['sstart'] - 4000\n",
    "        save_fasta(p_filtered_first_bed[['sseqid', 'sstart', 'send']], f'first/p_normal_{o}.txt')\n",
    "    \n",
    "        m_filtered_first_bed['end'] = m_filtered_first_bed['send'] + 4000\n",
    "        m_filtered_first_bed['start'] = m_filtered_first_bed['sstart'] - 4000\n",
    "        save_fasta(m_filtered_first_bed[['sseqid', 'sstart', 'send']], f'first/m_normal_{o}.txt')\n",
    "    \n",
    "        save_fasta(p_filtered_first_bed[['sseqid']], f'first/p_seq_list_{o}.txt')\n",
    "        save_fasta(m_filtered_first_bed[['sseqid']], f'first/m_seq_list_{o}.txt')\n",
    "        print(\"second part\")\n",
    "    \n",
    "        #long\n",
    "        os.system(f\"seqtk subseq ont/{o}.fasta first/p_seq_list_{o}.txt > first/p_normal_{o}.fasta\")\n",
    "        os.system(f\"seqtk subseq ont/{o}.fasta first/m_seq_list_{o}.txt > first/m_normal_{o}.fasta\")\n",
    "        print(\"1 part\")\n",
    "\n",
    "        os.system(f\"bedtools maskfasta -fi first/p_normal_{o}.fasta -bed first/p_normal_{o}.txt -fo first/p_Alex_{o}.masked.fasta\")\n",
    "        os.system(f\"seqkit fx2tab --length --name  first/p_Alex_{o}.masked.fasta > first/p_max_lenght_{o}.txt\")\n",
    "        print(\"2 part\")\n",
    "        p_max_len = pd.read_csv(\n",
    "            f'first/p_max_lenght_{o}.txt', sep=\"\\t\", header=None).rename(\n",
    "            {0:'sseqid', 1:'max_len'}, axis=1).sort_values(by=['sseqid'])\n",
    "        #print(\"3 part\")\n",
    "        os.system(f\"bedtools maskfasta -fi first/m_normal_{o}.fasta -bed first/m_normal_{o}.txt -fo first/m_Alex_{o}.masked.fasta\")\n",
    "        os.system(f\"seqkit fx2tab --length --name  first/m_Alex_{o}.masked.fasta > first/m_max_lenght_{o}.txt\")\n",
    "        print(\"4 part\")\n",
    "    \n",
    "        m_max_len = pd.read_csv(\n",
    "            f'first/m_max_lenght_{o}.txt', sep=\"\\t\", header=None).rename(\n",
    "            {0:'sseqid', 1:'max_len'}, axis=1).sort_values(by=['sseqid'])\n",
    "    \n",
    "        #sseqid bo query to po prostu te w tym przypadku\n",
    "        save_fasta(p_filtered_first_bed[['sseqid', 'start', 'end']], 'p_cutted_{o}.txt')\n",
    "        save_fasta(m_filtered_first_bed[['sseqid', 'start', 'end']], 'm_cutted_{o}.txt')\n",
    "        print(\"third part\")\n",
    "\n",
    "        \"\"\" \"\"\"\n",
    "        p_mask_min = p_filtered_first_bed.start < 0\n",
    "        p_filtered_first_bed.loc[p_mask_min, 'start'] = 0\n",
    "        p_filtered_first_bed = pd.merge(p_filtered_first_bed, p_max_len, how='left', on=\"sseqid\")\n",
    "        p_mask_max = p_filtered_first_bed.end > p_filtered_first_bed.max_len\n",
    "        p_filtered_first_bed.loc[p_mask_max, 'end'] = p_filtered_first_bed.max_len\n",
    "        save_fasta(p_filtered_first_bed[['sseqid', 'start', 'end']], f'first/p_cutted2_{o}.txt')\n",
    "        print(\"forth part\")\n",
    "        \"\"\" \"\"\"\n",
    "    \n",
    "        m_mask_min = m_filtered_first_bed.start < 0\n",
    "        m_filtered_first_bed.loc[m_mask_min, 'start'] = 0\n",
    "        m_filtered_first_bed = pd.merge(m_filtered_first_bed, m_max_len, how='left', on=\"sseqid\")\n",
    "        m_mask_max = m_filtered_first_bed.end > m_filtered_first_bed.max_len\n",
    "        m_filtered_first_bed.loc[m_mask_max, 'end'] = m_filtered_first_bed.max_len\n",
    "        save_fasta(m_filtered_first_bed[['sseqid', 'start', 'end']], f'first/m_cutted2_{o}.txt')\n",
    "    \n",
    "        # bedtools getfasta\n",
    "        os.system(f\"seqtk subseq first/p_Alex_{o}.masked.fasta first/p_cutted2_{o}.txt > first/p_cutted_ont_{o}.fasta\")\n",
    "        os.system(f\"seqtk subseq first/m_Alex_{o}.masked.fasta first/m_cutted2_{o}.txt > first/m_cutted_ont_{o}.fasta\")\n",
    "        print(\"fifth part\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "ee0890a8-71d0-456b-b474-31949b72f893",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 04/17/2024 05:44:24\n",
      "New DB name:   /home/kiki/data2/ont/konik4\n",
      "New DB title:  ont/konik4.fasta\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 44629 sequences in 13.0739 seconds.\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "first part\n",
      "second part\n",
      "1 part\n",
      "2 part\n",
      "4 part\n",
      "third part\n",
      "forth part\n",
      "fifth part\n"
     ]
    }
   ],
   "source": [
    "first_part()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "2faf8ba2-e5e0-4f26-a31f-beaf5e5a9c35",
   "metadata": {},
   "outputs": [],
   "source": [
    "h = [0, \"from\", \"to\"]\n",
    "save_fasta(doxi[h], 'doxi')\n",
    "!seqtk subseq genome/masked_DCARv3.4.fa doxi > dixi.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "b153eec7-4a7b-4e93-934d-3811d44c80bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 04/08/2024 05:52:47\n",
      "New DB name:   /home/kiki/data/okno/okno\n",
      "New DB title:  ont/K10B.fasta\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 6939343 sequences in 341.898 seconds.\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "!mkdir okno\n",
    "os.system(\"makeblastdb -in ont/K10B.fasta -dbtype nucl -out okno/okno\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "3de8604d-a84e-4f35-bbee-b32ff4768325",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.system(\"blastn -num_threads 20 -outfmt 6 -query dixi.fasta -db okno/okno -out wuli -perc_identity 0.9\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c1005ca-6ede-46f2-944b-305620683911",
   "metadata": {},
   "source": [
    "# Second part of analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "6e06d225-049b-43e1-8dfe-599b5ad2b05d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 04/17/2024 22:23:31\n",
      "New DB name:   /home/kiki/data2/genome/masked_genome\n",
      "New DB title:  genome/genomic2.fna\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /home/kiki/data2/genome/masked_genome\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 9 sequences in 4.86382 seconds.\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n",
      "Warning: [blastn] Number of threads was reduced to 8 to match the number of available CPUs\n"
     ]
    }
   ],
   "source": [
    "def second_part(blast_db=False):\n",
    "    all_report = []\n",
    "    os.system(\"mkdir -p final\")\n",
    "    os.system(\"mkdir -p second\")\n",
    "    pd.options.mode.copy_on_write = True\n",
    "    if blast_db:\n",
    "        # todo\n",
    "        os.system(f\"makeblastdb -in {genome_path}  -dbtype nucl -out genome/masked_genome\")\n",
    "    for o in ont:\n",
    "        #os.system(\"blastn -num_threads 20 -outfmt 6 -query p_cutted_ont.fasta -db genome/masked_DCARv3.4 -out p_Alex_K10B_genome.bl -perc_identity 0.8 -max_target_seqs 4\")\n",
    "        os.system(f\"blastn -num_threads 20 -outfmt 6 -query first/p_cutted_ont_{o}.fasta -db genome/masked_genome -out second/p_Alex_{o}_genome.bl -perc_identity 0.9\")\n",
    "        #os.system(\"blastn -num_threads 20 -outfmt 6 -query m_cutted_ont.fasta -db genome/masked_DCARv3.4 -out m_Alex_K10B_genome.bl -perc_identity 0.8 -max_target_seqs 4\")\n",
    "        os.system(f\"blastn -num_threads 20 -outfmt 6 -query first/m_cutted_ont_{o}.fasta -db genome/masked_genome -out second/m_Alex_{o}_genome.bl -perc_identity 0.9\")\n",
    "    \n",
    "        plus_blast = pd.read_csv(f\"second/p_Alex_{o}_genome.bl\", sep=\"\\t\", header=None).rename(headers_dict, axis=1)\n",
    "        minus_blast = pd.read_csv(f\"second/m_Alex_{o}_genome.bl\", sep=\"\\t\", header=None).rename(headers_dict, axis=1)\n",
    "    \n",
    "        p_plus_blast = plus_blast[plus_blast.sstart <= plus_blast.send]\n",
    "        m_plus_blast = plus_blast[plus_blast.sstart > plus_blast.send]\n",
    "        \n",
    "        p_minus_blast = minus_blast[minus_blast.sstart <= minus_blast.send]\n",
    "        m_minus_blast = minus_blast[minus_blast.sstart > minus_blast.send]\n",
    "    \n",
    "        p_plus = create_report(p_plus_blast)\n",
    "        m_plus = create_report(m_plus_blast)\n",
    "        p_minus = create_report(p_minus_blast)\n",
    "        m_minus = create_report(m_minus_blast)\n",
    "\n",
    "        #todo beter filter\n",
    "        p_plus = create_report(p_plus_blast[(p_plus_blast.length > 1500)])\n",
    "        m_plus = create_report(m_plus_blast[(m_plus_blast.length > 1500)])\n",
    "        p_minus = create_report(p_minus_blast[(p_minus_blast.length > 1500)])\n",
    "        m_minus = create_report(m_minus_blast[(m_minus_blast.length > 1500)])\n",
    "        \n",
    "        #p_plus.to_csv(f\"second/p_plus_report_{o}\", sep='\\t', index=False, encoding='utf-8', header=True) # i add header\n",
    "        #m_plus.to_csv(f\"second/m_plus_report_{o}\", sep='\\t', index=False, encoding='utf-8', header=True) # i add header\n",
    "        #p_minus.to_csv(f\"second/p_minus_report_{o}\", sep='\\t', index=False, encoding='utf-8', header=True) # i add header\n",
    "        #m_minus.to_csv(f\"second/m_minus_report_{o}\", sep='\\t', index=False, encoding='utf-8', header=True) # i add header\n",
    "    \n",
    "        p_plus_final = filter_finish(p_plus)\n",
    "        m_plus_final = filter_finish(m_plus)\n",
    "        p_minus_final = filter_finish(p_minus)\n",
    "        m_minus_final = filter_finish(m_minus)\n",
    "        \n",
    "        os.system(\"mkdir -p final\")\n",
    "        save_fasta(p_plus_final, f'final/p_plus_final_{o}')\n",
    "        save_fasta(m_plus_final, f'final/m_plus_final_{o}')\n",
    "        save_fasta(p_minus_final, f'final/p_minus_final_{o}')\n",
    "        save_fasta(m_minus_final, f'final/m_minus_final_{o}')\n",
    "        \n",
    "        final_concat =  pd.concat(\n",
    "            [\n",
    "                p_plus_final, \n",
    "                m_plus_final, \n",
    "                p_minus_final, \n",
    "                m_minus_final, \n",
    "            ]\n",
    "        ).sort_values(by=['sseqid', 'start_genome_te'])\n",
    "        all_report.append(final_concat)\n",
    "    \n",
    "        #fin[final_.end_genome_te - ni.start_genome_te > 20]\n",
    "        #return final_concat\n",
    "    return pd.concat(all_report)\n",
    "    \n",
    "oxi = second_part(True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "882b304b-5b98-4f96-9990-72b7ee53722b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>start_te</th>\n",
       "      <th>end_te</th>\n",
       "      <th>start_genome_te</th>\n",
       "      <th>end_genome_te</th>\n",
       "      <th>qstart1</th>\n",
       "      <th>qend1</th>\n",
       "      <th>sstart1</th>\n",
       "      <th>send1</th>\n",
       "      <th>qstart2</th>\n",
       "      <th>qend2</th>\n",
       "      <th>sstart2</th>\n",
       "      <th>send2</th>\n",
       "      <th>length</th>\n",
       "      <th>length_genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chromosome5_0_9528:3215-15939</td>\n",
       "      <td>chromosome5</td>\n",
       "      <td>4000</td>\n",
       "      <td>8726</td>\n",
       "      <td>8967214</td>\n",
       "      <td>8971940</td>\n",
       "      <td>8726</td>\n",
       "      <td>12725</td>\n",
       "      <td>8971940</td>\n",
       "      <td>8975939</td>\n",
       "      <td>2094</td>\n",
       "      <td>4000</td>\n",
       "      <td>8965308</td>\n",
       "      <td>8967214</td>\n",
       "      <td>4726</td>\n",
       "      <td>4726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chromosome5_1_7621:13215-25000</td>\n",
       "      <td>chromosome5</td>\n",
       "      <td>4000</td>\n",
       "      <td>8726</td>\n",
       "      <td>8967214</td>\n",
       "      <td>8971940</td>\n",
       "      <td>8726</td>\n",
       "      <td>11786</td>\n",
       "      <td>8971940</td>\n",
       "      <td>8975000</td>\n",
       "      <td>2094</td>\n",
       "      <td>4000</td>\n",
       "      <td>8965308</td>\n",
       "      <td>8967214</td>\n",
       "      <td>4726</td>\n",
       "      <td>4726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome9_0_16872:1-10148</td>\n",
       "      <td>chromosome9</td>\n",
       "      <td>466</td>\n",
       "      <td>6149</td>\n",
       "      <td>9220466</td>\n",
       "      <td>9226149</td>\n",
       "      <td>6149</td>\n",
       "      <td>10148</td>\n",
       "      <td>9226149</td>\n",
       "      <td>9230148</td>\n",
       "      <td>1</td>\n",
       "      <td>466</td>\n",
       "      <td>9220001</td>\n",
       "      <td>9220466</td>\n",
       "      <td>5683</td>\n",
       "      <td>5683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chromosome9_0_17837:5340-18568</td>\n",
       "      <td>chromosome9</td>\n",
       "      <td>3999</td>\n",
       "      <td>9230</td>\n",
       "      <td>28345736</td>\n",
       "      <td>28345740</td>\n",
       "      <td>9230</td>\n",
       "      <td>9437</td>\n",
       "      <td>28345740</td>\n",
       "      <td>28345533</td>\n",
       "      <td>3923</td>\n",
       "      <td>3999</td>\n",
       "      <td>28345810</td>\n",
       "      <td>28345736</td>\n",
       "      <td>5231</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chromosome9_1_14269:340-13568</td>\n",
       "      <td>chromosome9</td>\n",
       "      <td>3999</td>\n",
       "      <td>9230</td>\n",
       "      <td>28345736</td>\n",
       "      <td>28345740</td>\n",
       "      <td>9230</td>\n",
       "      <td>9437</td>\n",
       "      <td>28345740</td>\n",
       "      <td>28345533</td>\n",
       "      <td>3923</td>\n",
       "      <td>3999</td>\n",
       "      <td>28345810</td>\n",
       "      <td>28345736</td>\n",
       "      <td>5231</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           qseqid       sseqid  start_te  end_te  \\\n",
       "1   chromosome5_0_9528:3215-15939  chromosome5      4000    8726   \n",
       "1  chromosome5_1_7621:13215-25000  chromosome5      4000    8726   \n",
       "0     chromosome9_0_16872:1-10148  chromosome9       466    6149   \n",
       "7  chromosome9_0_17837:5340-18568  chromosome9      3999    9230   \n",
       "7   chromosome9_1_14269:340-13568  chromosome9      3999    9230   \n",
       "\n",
       "   start_genome_te  end_genome_te  qstart1  qend1   sstart1     send1  \\\n",
       "1          8967214        8971940     8726  12725   8971940   8975939   \n",
       "1          8967214        8971940     8726  11786   8971940   8975000   \n",
       "0          9220466        9226149     6149  10148   9226149   9230148   \n",
       "7         28345736       28345740     9230   9437  28345740  28345533   \n",
       "7         28345736       28345740     9230   9437  28345740  28345533   \n",
       "\n",
       "   qstart2  qend2   sstart2     send2  length  length_genome  \n",
       "1     2094   4000   8965308   8967214    4726           4726  \n",
       "1     2094   4000   8965308   8967214    4726           4726  \n",
       "0        1    466   9220001   9220466    5683           5683  \n",
       "7     3923   3999  28345810  28345736    5231              4  \n",
       "7     3923   3999  28345810  28345736    5231              4  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oxi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "8d441e7b-8564-4e34-ac12-0d68af986026",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chromosome1': {18795334.0: [18795334.0, 18799059.0, 1]},\n",
       " 'chromosome2': {30539632.0: [30539632.0, 30544358.0, 1]},\n",
       " 'chromosome3': {27717880.0: [27717880.0, 27722607.0, 1],\n",
       "  41159456.0: [41159456.0, 41164183.0, 1]},\n",
       " 'chromosome4': {35827649.0: [35827649.0, 35832375.0, 3]},\n",
       " 'chromosome5': {8967214.0: [8967214.0, 8971940.0, 2]},\n",
       " 'chromosome9': {9220466.0: [9220466.0, 9226149.0, 1],\n",
       "  28529339.0: [28529339.0, 28534569.0, 2]}}"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def mid(rec, row):\n",
    "    start =  row['start_genome_te']\n",
    "    for m in rec.keys():\n",
    "        if start - m < 1000 and start - m > -1000:\n",
    "            rec[m][2] =  rec[m][2] + 1\n",
    "            return\n",
    "    rec[start] =  [row['start_genome_te'], row['end_genome_te'], 1]\n",
    "    return\n",
    "    \n",
    "\n",
    "records = {}\n",
    "for sid in oxi.sseqid.unique():\n",
    "    okno = oxi[(oxi.sseqid == sid)]\n",
    "    records[sid] = {}\n",
    "    for index, row in okno.iterrows():\n",
    "        mid(records[sid], row) \n",
    "\n",
    "\n",
    "records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "029f1685-b48c-4ff9-b127-5bfe998bf627",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>start_te</th>\n",
       "      <th>end_te</th>\n",
       "      <th>start_genome_te</th>\n",
       "      <th>end_genome_te</th>\n",
       "      <th>qstart1</th>\n",
       "      <th>qend1</th>\n",
       "      <th>sstart1</th>\n",
       "      <th>send1</th>\n",
       "      <th>qstart2</th>\n",
       "      <th>qend2</th>\n",
       "      <th>sstart2</th>\n",
       "      <th>send2</th>\n",
       "      <th>length</th>\n",
       "      <th>length_genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9676ad26-b838-4bf4-b6f3-803f3cb06ff0:8506-20811</td>\n",
       "      <td>chromosome1</td>\n",
       "      <td>4000</td>\n",
       "      <td>8307</td>\n",
       "      <td>1895047</td>\n",
       "      <td>1895034</td>\n",
       "      <td>8307</td>\n",
       "      <td>9048</td>\n",
       "      <td>1895034</td>\n",
       "      <td>1894286</td>\n",
       "      <td>1</td>\n",
       "      <td>4000</td>\n",
       "      <td>1899099</td>\n",
       "      <td>1895047</td>\n",
       "      <td>4307</td>\n",
       "      <td>-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2c40afb0-7c90-4ddf-847c-a587e291d4a9:1-9021</td>\n",
       "      <td>chromosome1</td>\n",
       "      <td>612</td>\n",
       "      <td>5042</td>\n",
       "      <td>1895047</td>\n",
       "      <td>1895016</td>\n",
       "      <td>5042</td>\n",
       "      <td>9021</td>\n",
       "      <td>1895016</td>\n",
       "      <td>1890954</td>\n",
       "      <td>31</td>\n",
       "      <td>612</td>\n",
       "      <td>1895639</td>\n",
       "      <td>1895047</td>\n",
       "      <td>4430</td>\n",
       "      <td>-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>08051692-8a2e-4ada-ae64-5ded5fdb190a:28450-38629</td>\n",
       "      <td>chromosome1</td>\n",
       "      <td>3821</td>\n",
       "      <td>8404</td>\n",
       "      <td>12813067</td>\n",
       "      <td>12813069</td>\n",
       "      <td>8404</td>\n",
       "      <td>9604</td>\n",
       "      <td>12813069</td>\n",
       "      <td>12814278</td>\n",
       "      <td>2697</td>\n",
       "      <td>3821</td>\n",
       "      <td>12811946</td>\n",
       "      <td>12813067</td>\n",
       "      <td>4583</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9ebca745-06aa-4187-9303-3fc73853bf74:1-10655</td>\n",
       "      <td>chromosome1</td>\n",
       "      <td>2054</td>\n",
       "      <td>6658</td>\n",
       "      <td>27115104</td>\n",
       "      <td>27115102</td>\n",
       "      <td>6658</td>\n",
       "      <td>10655</td>\n",
       "      <td>27115102</td>\n",
       "      <td>27111051</td>\n",
       "      <td>27</td>\n",
       "      <td>2054</td>\n",
       "      <td>27117184</td>\n",
       "      <td>27115104</td>\n",
       "      <td>4604</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>98c0c670-f831-449a-a2d7-6b50aad7d341:308-12808</td>\n",
       "      <td>chromosome1</td>\n",
       "      <td>4000</td>\n",
       "      <td>8546</td>\n",
       "      <td>40590593</td>\n",
       "      <td>40590596</td>\n",
       "      <td>8546</td>\n",
       "      <td>10780</td>\n",
       "      <td>40590596</td>\n",
       "      <td>40588308</td>\n",
       "      <td>1173</td>\n",
       "      <td>4000</td>\n",
       "      <td>40593451</td>\n",
       "      <td>40590593</td>\n",
       "      <td>4546</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             qseqid       sseqid  start_te  \\\n",
       "0   9676ad26-b838-4bf4-b6f3-803f3cb06ff0:8506-20811  chromosome1      4000   \n",
       "0       2c40afb0-7c90-4ddf-847c-a587e291d4a9:1-9021  chromosome1       612   \n",
       "9  08051692-8a2e-4ada-ae64-5ded5fdb190a:28450-38629  chromosome1      3821   \n",
       "0      9ebca745-06aa-4187-9303-3fc73853bf74:1-10655  chromosome1      2054   \n",
       "2    98c0c670-f831-449a-a2d7-6b50aad7d341:308-12808  chromosome1      4000   \n",
       "\n",
       "   end_te  start_genome_te  end_genome_te  qstart1  qend1   sstart1     send1  \\\n",
       "0    8307          1895047        1895034     8307   9048   1895034   1894286   \n",
       "0    5042          1895047        1895016     5042   9021   1895016   1890954   \n",
       "9    8404         12813067       12813069     8404   9604  12813069  12814278   \n",
       "0    6658         27115104       27115102     6658  10655  27115102  27111051   \n",
       "2    8546         40590593       40590596     8546  10780  40590596  40588308   \n",
       "\n",
       "   qstart2  qend2   sstart2     send2  length  length_genome  \n",
       "0        1   4000   1899099   1895047    4307            -13  \n",
       "0       31    612   1895639   1895047    4430            -31  \n",
       "9     2697   3821  12811946  12813067    4583              2  \n",
       "0       27   2054  27117184  27115104    4604             -2  \n",
       "2     1173   4000  40593451  40590593    4546              3  "
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oxi.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf8fa90c-b11d-4a68-9e9e-b361f5f5b7d9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
